Thank you for visiting nature.com. You are using a browser version with limited support for CSS. To obtain the best experience, we recommend you use a more up to date browser (or turn off compatibility mode in Internet Explorer). In the meantime, to ensure continued support, we are displaying the site without styles and JavaScript.
Advertisement
[ ![Nature](https://media.springernature.com/full/nature-cms/uploads/product/nature/header-86f1267ea01eccd46b530284be10585e.svg) ](https://www.nature.com/)
  * [ View all journals ](https://www.nature.com/siteindex)
  * [ Search ](javascript:;)
## Search
Search articles by subject, keyword or author
Show results from All journals This journal
Search
[ Advanced search ](https://www.nature.com/search/advanced)
### Quick links
    * [Explore articles by subject](https://www.nature.com/subjects)
    * [Find a job](https://www.nature.com/naturecareers)
    * [Guide to authors](https://www.nature.com/authors/index.html)
    * [Editorial policies](https://www.nature.com/authors/editorial_policies/)
  * [Log in](https://idp.nature.com/auth/personal/springernature?redirect_uri=https://www.nature.com/articles/s41586-020-2649-2)


  * [ Content Explore content](javascript:;)
## Explore content
    * [ Research articles ](https://www.nature.com/nature/research-articles)
    * [ News ](https://www.nature.com/news)
    * [ Opinion ](https://www.nature.com/opinion)
    * [ Research Analysis ](https://www.nature.com/research-analysis)
    * [ Careers ](https://www.nature.com/careers)
    * [ Books & Culture ](https://www.nature.com/books-culture)
    * [ Podcasts ](https://www.nature.com/nature/podcasts)
    * [ Videos ](https://www.nature.com/nature/videos)
    * [ Current issue ](https://www.nature.com/nature/current-issue)
    * [ Browse issues ](https://www.nature.com/nature/browse-issues)
    * [ Collections ](https://www.nature.com/nature/collections)
    * [ Subjects ](https://www.nature.com/nature/browse-subjects)
    * [Follow us on Facebook ](https://www.facebook.com/Nature)
    * [Follow us on Twitter ](https://twitter.com/nature)
    * [Sign up for alerts](https://journal-alerts.springernature.com/subscribe?journal_id=41586)
    * [ RSS feed ](https://www.nature.com/nature.rss)
  * [ About the journal](javascript:;)
## About the journal
    * [ Journal Staff ](https://www.nature.com/nature/journal-staff)
    * [ About the Editors ](https://www.nature.com/nature/editors)
    * [ Journal Information ](https://www.nature.com/nature/journal-information)
    * [ Journal Metrics ](https://www.nature.com/nature/journal-impact)
    * [ Our publishing models ](https://www.nature.com/nature/our-publishing-models)
    * [ Editorial Values Statement ](https://www.nature.com/nature/editorial-values-statement)
    * [ Editorial policies ](https://www.nature.com/nature/editorial-policies)
    * [ Journalistic Principles ](https://www.nature.com/nature/journalistic-principles)
    * [ History of Nature ](https://www.nature.com/nature/history-of-nature)
    * [ Awards ](https://www.nature.com/nature/awards)
    * [ Contact ](https://www.nature.com/nature/contact)
    * [ Send a news tip ](https://www.nature.com/nature/send-a-news-tip)
  * [ Publish with us](javascript:;)
## Publish with us
    * [ For Authors ](https://www.nature.com/nature/for-authors)
    * [ For Referees ](https://www.nature.com/nature/for-referees)
    * [ Language editing services ](https://authorservices.springernature.com/go/sn/?utm_source=For+Authors&utm_medium=Website_Nature&utm_campaign=Platform+Experimentation+2022&utm_id=PE2022)
    * [Open access funding](https://www.nature.com/nature/open-access-funding)
    * [Submit manuscript](https://mts-nature.nature.com/)


  * [ Sign up for alerts](https://journal-alerts.springernature.com/subscribe?journal_id=41586)
  * [ RSS feed ](https://www.nature.com/nature.rss)


  1. [nature](https://www.nature.com/)
  2. [review articles](https://www.nature.com/nature/articles?type=review-article)
  3. article


Array programming with NumPy 
[ Download PDF ](https://www.nature.com/articles/s41586-020-2649-2.pdf)
[ Download PDF ](https://www.nature.com/articles/s41586-020-2649-2.pdf)
  * Review Article
  * [Open access](https://www.springernature.com/gp/open-science/about/the-fundamentals-of-open-access-and-open-research)
  * Published: 16 September 2020


# Array programming with NumPy
  * [Charles R. Harris](https://www.nature.com/articles/s41586-020-2649-2#auth-Charles_R_-Harris-Aff1)[1](https://www.nature.com/articles/s41586-020-2649-2#Aff1), 
  * [K. Jarrod Millman](https://www.nature.com/articles/s41586-020-2649-2#auth-K__Jarrod-Millman-Aff2-Aff3-Aff4) [ORCID: orcid.org/0000-0002-5263-5070](https://orcid.org/0000-0002-5263-5070)[2](https://www.nature.com/articles/s41586-020-2649-2#Aff2),[3](https://www.nature.com/articles/s41586-020-2649-2#Aff3),[4](https://www.nature.com/articles/s41586-020-2649-2#Aff4), 
  * [Stéfan J. van der Walt](https://www.nature.com/articles/s41586-020-2649-2#auth-St_fan_J_-Walt-Aff2-Aff4-Aff5) [ORCID: orcid.org/0000-0001-9276-1891](https://orcid.org/0000-0001-9276-1891)[2](https://www.nature.com/articles/s41586-020-2649-2#Aff2),[4](https://www.nature.com/articles/s41586-020-2649-2#Aff4),[5](https://www.nature.com/articles/s41586-020-2649-2#Aff5), 
  * [Ralf Gommers](https://www.nature.com/articles/s41586-020-2649-2#auth-Ralf-Gommers-Aff6) [ORCID: orcid.org/0000-0002-0300-3333](https://orcid.org/0000-0002-0300-3333)[6](https://www.nature.com/articles/s41586-020-2649-2#Aff6), 
  * [Pauli Virtanen](https://www.nature.com/articles/s41586-020-2649-2#auth-Pauli-Virtanen-Aff7-Aff8)[7](https://www.nature.com/articles/s41586-020-2649-2#Aff7),[8](https://www.nature.com/articles/s41586-020-2649-2#Aff8), 
  * [David Cournapeau](https://www.nature.com/articles/s41586-020-2649-2#auth-David-Cournapeau-Aff9)[9](https://www.nature.com/articles/s41586-020-2649-2#Aff9), 
  * [Eric Wieser](https://www.nature.com/articles/s41586-020-2649-2#auth-Eric-Wieser-Aff10)[10](https://www.nature.com/articles/s41586-020-2649-2#Aff10), 
  * [Julian Taylor](https://www.nature.com/articles/s41586-020-2649-2#auth-Julian-Taylor-Aff11)[11](https://www.nature.com/articles/s41586-020-2649-2#Aff11), 
  * [Sebastian Berg](https://www.nature.com/articles/s41586-020-2649-2#auth-Sebastian-Berg-Aff4)[4](https://www.nature.com/articles/s41586-020-2649-2#Aff4), 
  * [Nathaniel J. Smith](https://www.nature.com/articles/s41586-020-2649-2#auth-Nathaniel_J_-Smith-Aff12)[12](https://www.nature.com/articles/s41586-020-2649-2#Aff12), 
  * [Robert Kern](https://www.nature.com/articles/s41586-020-2649-2#auth-Robert-Kern-Aff13)[13](https://www.nature.com/articles/s41586-020-2649-2#Aff13), 
  * [Matti Picus](https://www.nature.com/articles/s41586-020-2649-2#auth-Matti-Picus-Aff4) [ORCID: orcid.org/0000-0002-1771-9949](https://orcid.org/0000-0002-1771-9949)[4](https://www.nature.com/articles/s41586-020-2649-2#Aff4), 
  * [Stephan Hoyer](https://www.nature.com/articles/s41586-020-2649-2#auth-Stephan-Hoyer-Aff14) [ORCID: orcid.org/0000-0002-5207-0380](https://orcid.org/0000-0002-5207-0380)[14](https://www.nature.com/articles/s41586-020-2649-2#Aff14), 
  * [Marten H. van Kerkwijk](https://www.nature.com/articles/s41586-020-2649-2#auth-Marten_H_-Kerkwijk-Aff15)[15](https://www.nature.com/articles/s41586-020-2649-2#Aff15), 
  * [Matthew Brett](https://www.nature.com/articles/s41586-020-2649-2#auth-Matthew-Brett-Aff2-Aff16)[2](https://www.nature.com/articles/s41586-020-2649-2#Aff2),[16](https://www.nature.com/articles/s41586-020-2649-2#Aff16), 
  * [Allan Haldane](https://www.nature.com/articles/s41586-020-2649-2#auth-Allan-Haldane-Aff17)[17](https://www.nature.com/articles/s41586-020-2649-2#Aff17), 
  * [Jaime Fernández del Río](https://www.nature.com/articles/s41586-020-2649-2#auth-Jaime_Fern_ndez-R_o-Aff18)[18](https://www.nature.com/articles/s41586-020-2649-2#Aff18), 
  * [Mark Wiebe](https://www.nature.com/articles/s41586-020-2649-2#auth-Mark-Wiebe-Aff19-Aff20) [ORCID: orcid.org/0000-0003-3603-8038](https://orcid.org/0000-0003-3603-8038)[19](https://www.nature.com/articles/s41586-020-2649-2#Aff19),[20](https://www.nature.com/articles/s41586-020-2649-2#Aff20), 
  * [Pearu Peterson](https://www.nature.com/articles/s41586-020-2649-2#auth-Pearu-Peterson-Aff6-Aff21-Aff22) [ORCID: orcid.org/0000-0001-7328-4305](https://orcid.org/0000-0001-7328-4305)[6](https://www.nature.com/articles/s41586-020-2649-2#Aff6),[21](https://www.nature.com/articles/s41586-020-2649-2#Aff21),[22](https://www.nature.com/articles/s41586-020-2649-2#Aff22), 
  * [Pierre Gérard-Marchant](https://www.nature.com/articles/s41586-020-2649-2#auth-Pierre-G_rard_Marchant-Aff23-Aff24)[23](https://www.nature.com/articles/s41586-020-2649-2#Aff23),[24](https://www.nature.com/articles/s41586-020-2649-2#Aff24), 
  * [Kevin Sheppard](https://www.nature.com/articles/s41586-020-2649-2#auth-Kevin-Sheppard-Aff25) [ORCID: orcid.org/0000-0001-8700-2292](https://orcid.org/0000-0001-8700-2292)[25](https://www.nature.com/articles/s41586-020-2649-2#Aff25), 
  * [Tyler Reddy](https://www.nature.com/articles/s41586-020-2649-2#auth-Tyler-Reddy-Aff26)[26](https://www.nature.com/articles/s41586-020-2649-2#Aff26), 
  * [Warren Weckesser](https://www.nature.com/articles/s41586-020-2649-2#auth-Warren-Weckesser-Aff4)[4](https://www.nature.com/articles/s41586-020-2649-2#Aff4), 
  * [Hameer Abbasi](https://www.nature.com/articles/s41586-020-2649-2#auth-Hameer-Abbasi-Aff6)[6](https://www.nature.com/articles/s41586-020-2649-2#Aff6), 
  * [Christoph Gohlke](https://www.nature.com/articles/s41586-020-2649-2#auth-Christoph-Gohlke-Aff27) [ORCID: orcid.org/0000-0001-8108-7707](https://orcid.org/0000-0001-8108-7707)[27](https://www.nature.com/articles/s41586-020-2649-2#Aff27) &
  * …
  * [Travis E. Oliphant](https://www.nature.com/articles/s41586-020-2649-2#auth-Travis_E_-Oliphant-Aff6)[6](https://www.nature.com/articles/s41586-020-2649-2#Aff6)

Show authors
[_Nature_](https://www.nature.com/) **585** , 357–362 (2020)[Cite this article](https://www.nature.com/articles/s41586-020-2649-2#citeas)
  * 498k Accesses
  * 23k Citations
  * 1634 Altmetric
  * [Metrics ](https://www.nature.com/articles/s41586-020-2649-2/metrics)


## Abstract
Array programming provides a powerful, compact and expressive syntax for accessing, manipulating and operating on data in vectors, matrices and higher-dimensional arrays. NumPy is the primary array programming library for the Python language. It has an essential role in research analysis pipelines in fields as diverse as physics, chemistry, astronomy, geoscience, biology, psychology, materials science, engineering, finance and economics. For example, in astronomy, NumPy was an important part of the software stack used in the discovery of gravitational waves[1](https://www.nature.com/articles/s41586-020-2649-2#ref-CR1 "Abbott, B. P. et al. Observation of gravitational waves from a binary black hole merger. Phys. Rev. Lett. 116, 061102 \(2016\).") and in the first imaging of a black hole[2](https://www.nature.com/articles/s41586-020-2649-2#ref-CR2 "Chael, A. et al. High-resolution linear polarimetric imaging for the Event Horizon Telescope. Astrophys. J. 286, 11 \(2016\)."). Here we review how a few fundamental array concepts lead to a simple and powerful programming paradigm for organizing, exploring and analysing scientific data. NumPy is the foundation upon which the scientific Python ecosystem is constructed. It is so pervasive that several projects, targeting audiences with specialized needs, have developed their own NumPy-like interfaces and array objects. Owing to its central position in the ecosystem, NumPy increasingly acts as an interoperability layer between such array computation libraries and, together with its application programming interface (API), provides a flexible framework to support the next decade of scientific and industrial analysis.
### Similar content being viewed by others
![](https://media.springernature.com/w215h120/springer-static/image/art%3A10.1038%2Fs41467-023-43932-6/MediaObjects/41467_2023_43932_Fig1_HTML.png)
###  [An updated nuclear-physics and multi-messenger astrophysics framework for binary neutron star mergers ](https://www.nature.com/articles/s41467-023-43932-6?fromPaywallRec=false)
Article Open access 20 December 2023
![](https://media.springernature.com/w215h120/springer-static/image/art%3A10.1038%2Fs41598-022-16547-y/MediaObjects/41598_2022_16547_Fig1_HTML.png)
###  [Deep null fixing on optimal compromise among sum and difference patterns of thinned arrays through the Schelkunoff unit circle representation ](https://www.nature.com/articles/s41598-022-16547-y?fromPaywallRec=false)
Article Open access 16 July 2022
![](https://media.springernature.com/w215h120/springer-static/image/art%3A10.1038%2Fs41586-022-04481-y/MediaObjects/41586_2022_4481_Fig1_HTML.png)
###  [X-ray astronomy comes of age ](https://www.nature.com/articles/s41586-022-04481-y?fromPaywallRec=false)
Article 08 June 2022
## Main
Two Python array packages existed before NumPy. The Numeric package was developed in the mid-1990s and provided array objects and array-aware functions in Python. It was written in C and linked to standard fast implementations of linear algebra[3](https://www.nature.com/articles/s41586-020-2649-2#ref-CR3 "Dubois, P. F., Hinsen, K. & Hugunin, J. Numerical Python. Comput. Phys. 10, 262–267 \(1996\)."),[4](https://www.nature.com/articles/s41586-020-2649-2#ref-CR4 "Ascher, D., Dubois, P. F., Hinsen, K., Hugunin, J. & Oliphant, T. E. An Open Source Project: Numerical Python \(Lawrence Livermore National Laboratory, 2001\)."). One of its earliest uses was to steer C++ applications for inertial confinement fusion research at Lawrence Livermore National Laboratory[5](https://www.nature.com/articles/s41586-020-2649-2#ref-CR5 "Yang, T.-Y., Furnish, G. & Dubois, P. F. Steering object-oriented scientific computations. In Proc. TOOLS USA 97. Intl Conf. Technology of Object Oriented Systems and Languages \(eds Ege, R., Singh, M. & Meyer, B.\) 112–119 \(IEEE, 1997\)."). To handle large astronomical images coming from the Hubble Space Telescope, a reimplementation of Numeric, called Numarray, added support for structured arrays, flexible indexing, memory mapping, byte-order variants, more efficient memory use, flexible IEEE 754-standard error-handling capabilities, and better type-casting rules[6](https://www.nature.com/articles/s41586-020-2649-2#ref-CR6 "Greenfield, P., Miller, J. T., Hsu, J. & White, R. L. numarray: a new scientific array package for Python. In PyCon DC 2003 
                  http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.112.9899
                  
                 \(2003\)."). Although Numarray was highly compatible with Numeric, the two packages had enough differences that it divided the community; however, in 2005 NumPy emerged as a ‘best of both worlds’ unification[7](https://www.nature.com/articles/s41586-020-2649-2#ref-CR7 "Oliphant, T. E. Guide to NumPy 1st edn \(Trelgol Publishing, 2006\).")—combining the features of Numarray with the small-array performance of Numeric and its rich C API.
Now, 15 years later, NumPy underpins almost every Python library that does scientific or numerical computation[8](https://www.nature.com/articles/s41586-020-2649-2#ref-CR8 "Dubois, P. F. Python: batteries included. Comput. Sci. Eng. 9, 7–9 \(2007\)."),[9](https://www.nature.com/articles/s41586-020-2649-2#ref-CR9 "Oliphant, T. E. Python for scientific computing. Comput. Sci. Eng. 9, 10–20 \(2007\)."),[10](https://www.nature.com/articles/s41586-020-2649-2#ref-CR10 "Millman, K. J. & Aivazis, M. Python for scientists and engineers. Comput. Sci. Eng. 13, 9–12 \(2011\)."),[11](https://www.nature.com/articles/s41586-020-2649-2#ref-CR11 "Pérez, F., Granger, B. E. & Hunter, J. D. Python: an ecosystem for scientific computing. Comput. Sci. Eng. 13, 13–21 \(2011\). Explains why the scientific Python ecosystem is a highly productive environment for research."), including SciPy[12](https://www.nature.com/articles/s41586-020-2649-2#ref-CR12 "Virtanen, P. et al. SciPy 1.0—fundamental algorithms for scientific computing in Python. Nat. Methods 17, 261–272 \(2020\); correction 17, 352 \(2020\). Introduces the SciPy library and includes a more detailed history of NumPy and SciPy.
                        "), Matplotlib[13](https://www.nature.com/articles/s41586-020-2649-2#ref-CR13 "Hunter, J. D. Matplotlib: a 2D graphics environment. Comput. Sci. Eng. 9, 90–95 \(2007\)."), pandas[14](https://www.nature.com/articles/s41586-020-2649-2#ref-CR14 "McKinney, W. Data structures for statistical computing in Python. In Proc. 9th Python in Science Conf. \(eds van der Walt, S. & Millman, K. J.\) 56–61 \(2010\)."), scikit-learn[15](https://www.nature.com/articles/s41586-020-2649-2#ref-CR15 "Pedregosa, F. et al. Scikit-learn: machine learning in Python. J. Mach. Learn. Res. 12, 2825–2830 \(2011\).") and scikit-image[16](https://www.nature.com/articles/s41586-020-2649-2#ref-CR16 "van der Walt, S. et al. scikit-image: image processing in Python. PeerJ 2, e453 \(2014\)."). NumPy is a community-developed, open-source library, which provides a multidimensional Python array object along with array-aware functions that operate on it. Because of its inherent simplicity, the NumPy array is the de facto exchange format for array data in Python.
NumPy operates on in-memory arrays using the central processing unit (CPU). To utilize modern, specialized storage and hardware, there has been a recent proliferation of Python array packages. Unlike with the Numarray–Numeric divide, it is now much harder for these new libraries to fracture the user community—given how much work is already built on top of NumPy. However, to provide the community with access to new and exploratory technologies, NumPy is transitioning into a central coordinating mechanism that specifies a well defined array programming API and dispatches it, as appropriate, to specialized array implementations.
## NumPy arrays
The NumPy array is a data structure that efficiently stores and accesses multidimensional arrays[17](https://www.nature.com/articles/s41586-020-2649-2#ref-CR17 "van der Walt, S., Colbert, S. C. & Varoquaux, G. The NumPy array: a structure for efficient numerical computation. Comput. Sci. Eng. 13, 22–30 \(2011\). Discusses the NumPy array data structure with a focus on how it enables efficient computation.") (also known as tensors), and enables a wide variety of scientific computation. It consists of a pointer to memory, along with metadata used to interpret the data stored there, notably ‘data type’, ‘shape’ and ‘strides’ (Fig. [1a](https://www.nature.com/articles/s41586-020-2649-2#Fig1)).
**Fig. 1: The NumPy array incorporates several fundamental array concepts.**
[![figure 1](https://media.springernature.com/lw685/springer-static/image/art%3A10.1038%2Fs41586-020-2649-2/MediaObjects/41586_2020_2649_Fig1_HTML.png)](https://www.nature.com/articles/s41586-020-2649-2/figures/1)
**a** , The NumPy array data structure and its associated metadata fields. **b** , Indexing an array with slices and steps. These operations return a ‘view’ of the original data. **c** , Indexing an array with masks, scalar coordinates or other arrays, so that it returns a ‘copy’ of the original data. In the bottom example, an array is indexed with other arrays; this broadcasts the indexing arguments before performing the lookup. **d** , Vectorization efficiently applies operations to groups of elements. **e** , Broadcasting in the multiplication of two-dimensional arrays. **f** , Reduction operations act along one or more axes. In this example, an array is summed along select axes to produce a vector, or along two axes consecutively to produce a scalar. **g** , Example NumPy code, illustrating some of these concepts.
[Full size image](https://www.nature.com/articles/s41586-020-2649-2/figures/1)
The data type describes the nature of elements stored in an array. An array has a single data type, and each element of an array occupies the same number of bytes in memory. Examples of data types include real and complex numbers (of lower and higher precision), strings, timestamps and pointers to Python objects.
The shape of an array determines the number of elements along each axis, and the number of axes is the dimensionality of the array. For example, a vector of numbers can be stored as a one-dimensional array of shape _N_ , whereas colour videos are four-dimensional arrays of shape (_T_ ,  _M_ ,  _N_ , 3).
Strides are necessary to interpret computer memory, which stores elements linearly, as multidimensional arrays. They describe the number of bytes to move forward in memory to jump from row to row, column to column, and so forth. Consider, for example, a two-dimensional array of floating-point numbers with shape (4, 3), where each element occupies 8 bytes in memory. To move between consecutive columns, we need to jump forward 8 bytes in memory, and to access the next row, 3 × 8 = 24 bytes. The strides of that array are therefore (24, 8). NumPy can store arrays in either C or Fortran memory order, iterating first over either rows or columns. This allows external libraries written in those languages to access NumPy array data in memory directly.
Users interact with NumPy arrays using ‘indexing’ (to access subarrays or individual elements), ‘operators’ (for example, +, − and × for vectorized operations and @ for matrix multiplication), as well as ‘array-aware functions’; together, these provide an easily readable, expressive, high-level API for array programming while NumPy deals with the underlying mechanics of making operations fast.
Indexing an array returns single elements, subarrays or elements that satisfy a specific condition (Fig. [1b](https://www.nature.com/articles/s41586-020-2649-2#Fig1)). Arrays can even be indexed using other arrays (Fig. [1c](https://www.nature.com/articles/s41586-020-2649-2#Fig1)). Wherever possible, indexing that retrieves a subarray returns a ‘view’ on the original array such that data are shared between the two arrays. This provides a powerful way to operate on subsets of array data while limiting memory usage.
To complement the array syntax, NumPy includes functions that perform vectorized calculations on arrays, including arithmetic, statistics and trigonometry (Fig. [1d](https://www.nature.com/articles/s41586-020-2649-2#Fig1)). Vectorization—operating on entire arrays rather than their individual elements—is essential to array programming. This means that operations that would take many tens of lines to express in languages such as C can often be implemented as a single, clear Python expression. This results in concise code and frees users to focus on the details of their analysis, while NumPy handles looping over array elements near-optimally—for example, taking strides into consideration to best utilize the computer’s fast cache memory.
When performing a vectorized operation (such as addition) on two arrays with the same shape, it is clear what should happen. Through ‘broadcasting’ NumPy allows the dimensions to differ, and produces results that appeal to intuition. A trivial example is the addition of a scalar value to an array, but broadcasting also generalizes to more complex examples such as scaling each column of an array or generating a grid of coordinates. In broadcasting, one or both arrays are virtually duplicated (that is, without copying any data in memory), so that the shapes of the operands match (Fig. [1d](https://www.nature.com/articles/s41586-020-2649-2#Fig1)). Broadcasting is also applied when an array is indexed using arrays of indices (Fig. [1c](https://www.nature.com/articles/s41586-020-2649-2#Fig1)).
Other array-aware functions, such as sum, mean and maximum, perform element-by-element ‘reductions’, aggregating results across one, multiple or all axes of a single array. For example, summing an _n_ -dimensional array over _d_ axes results in an array of dimension _n_ −  _d_ (Fig. [1f](https://www.nature.com/articles/s41586-020-2649-2#Fig1)).
NumPy also includes array-aware functions for creating, reshaping, concatenating and padding arrays; searching, sorting and counting data; and reading and writing files. It provides extensive support for generating pseudorandom numbers, includes an assortment of probability distributions, and performs accelerated linear algebra, using one of several backends such as OpenBLAS[18](https://www.nature.com/articles/s41586-020-2649-2#ref-CR18 "Wang, Q., Zhang, X., Zhang, Y. & Yi, Q. AUGEM: automatically generate high performance dense linear algebra kernels on x86 CPUs. In SC’13: Proc. Intl Conf. High Performance Computing, Networking, Storage and Analysis 25 \(IEEE, 2013\)."),[19](https://www.nature.com/articles/s41586-020-2649-2#ref-CR19 "Xianyi, Z., Qian, W. & Yunquan, Z. Model-driven level 3 BLAS performance optimization on Loongson 3A processor. In 2012 IEEE 18th Intl Conf. Parallel and Distributed Systems 684–691 \(IEEE, 2012\).") or Intel MKL optimized for the CPUs at hand (see Supplementary Methods for more details).
Altogether, the combination of a simple in-memory array representation, a syntax that closely mimics mathematics, and a variety of array-aware utility functions forms a productive and powerfully expressive array programming language.
## Scientific Python ecosystem
Python is an open-source, general-purpose interpreted programming language well suited to standard programming tasks such as cleaning data, interacting with web resources and parsing text. Adding fast array operations and linear algebra enables scientists to do all their work within a single programming language—one that has the advantage of being famously easy to learn and teach, as witnessed by its adoption as a primary learning language in many universities.
Even though NumPy is not part of Python’s standard library, it benefits from a good relationship with the Python developers. Over the years, the Python language has added new features and special syntax so that NumPy would have a more succinct and easier-to-read array notation. However, because it is not part of the standard library, NumPy is able to dictate its own release policies and development patterns.
SciPy and Matplotlib are tightly coupled with NumPy in terms of history, development and use. SciPy provides fundamental algorithms for scientific computing, including mathematical, scientific and engineering routines. Matplotlib generates publication-ready figures and visualizations. The combination of NumPy, SciPy and Matplotlib, together with an advanced interactive environment such as IPython[20](https://www.nature.com/articles/s41586-020-2649-2#ref-CR20 "Pérez, F. & Granger, B. E. IPython: a system for interactive scientific computing. Comput. Sci. Eng. 9, 21–29 \(2007\).") or Jupyter[21](https://www.nature.com/articles/s41586-020-2649-2#ref-CR21 "Kluyver, T. et al. Jupyter Notebooks—a publishing format for reproducible computational workflows. In Positioning and Power in Academic Publishing: Players, Agents and Agendas \(eds Loizides, F. & Schmidt, B.\) 87–90 \(IOS Press, 2016\)."), provides a solid foundation for array programming in Python. The scientific Python ecosystem (Fig. [2](https://www.nature.com/articles/s41586-020-2649-2#Fig2)) builds on top of this foundation to provide several, widely used technique-specific libraries[15](https://www.nature.com/articles/s41586-020-2649-2#ref-CR15 "Pedregosa, F. et al. Scikit-learn: machine learning in Python. J. Mach. Learn. Res. 12, 2825–2830 \(2011\)."),[16](https://www.nature.com/articles/s41586-020-2649-2#ref-CR16 "van der Walt, S. et al. scikit-image: image processing in Python. PeerJ 2, e453 \(2014\)."),[22](https://www.nature.com/articles/s41586-020-2649-2#ref-CR22 "Hagberg, A. A., Schult, D. A. & Swart, P. J. Exploring network structure, dynamics, and function using NetworkX. In Proc. 7th Python in Science Conf. \(eds Varoquaux, G., Vaught, T. & Millman, K. J.\) 11–15 \(2008\)."), that in turn underlie numerous domain-specific projects[23](https://www.nature.com/articles/s41586-020-2649-2#ref-CR23 "Astropy Collaboration et al. Astropy: a community Python package for astronomy. Astron. Astrophys. 558, A33 \(2013\)."),[24](https://www.nature.com/articles/s41586-020-2649-2#ref-CR24 "Price-Whelan, A. M. et al. The Astropy Project: building an open-science project and status of the v2.0 core package. Astron. J. 156, 123 \(2018\)."),[25](https://www.nature.com/articles/s41586-020-2649-2#ref-CR25 "Cock, P. J. et al. Biopython: freely available Python tools for computational molecular biology and bioinformatics. Bioinformatics 25, 1422–1423 \(2009\)."),[26](https://www.nature.com/articles/s41586-020-2649-2#ref-CR26 "Millman, K. J. & Brett, M. Analysis of functional magnetic resonance imaging in Python. Comput. Sci. Eng. 9, 52–55 \(2007\)."),[27](https://www.nature.com/articles/s41586-020-2649-2#ref-CR27 "The SunPy Community et al. SunPy—Python for solar physics. Comput. Sci. Discov. 8, 014009 \(2015\)."),[28](https://www.nature.com/articles/s41586-020-2649-2#ref-CR28 "Hamman, J., Rocklin, M. & Abernathy, R. Pangeo: a big-data ecosystem for scalable Earth system science. In EGU General Assembly Conf. Abstracts 12146 \(2018\)."). NumPy, at the base of the ecosystem of array-aware libraries, sets documentation standards, provides array testing infrastructure and adds build support for Fortran and other compilers.
**Fig. 2: NumPy is the base of the scientific Python ecosystem.**
[![figure 2](https://media.springernature.com/lw685/springer-static/image/art%3A10.1038%2Fs41586-020-2649-2/MediaObjects/41586_2020_2649_Fig2_HTML.png)](https://www.nature.com/articles/s41586-020-2649-2/figures/2)
Essential libraries and projects that depend on NumPy’s API gain access to new array implementations that support NumPy’s array protocols (Fig. [3](https://www.nature.com/articles/s41586-020-2649-2#Fig3)).
[Full size image](https://www.nature.com/articles/s41586-020-2649-2/figures/2)
Many research groups have designed large, complex scientific libraries that add application-specific functionality to the ecosystem. For example, the eht-imaging library[29](https://www.nature.com/articles/s41586-020-2649-2#ref-CR29 "Chael, A. A. et al. ehtim: imaging, analysis, and simulation software for radio interferometry. Astrophysics Source Code Library 
                  https://ascl.net/1904.004
                  
                 \(2019\)."), developed by the Event Horizon Telescope collaboration for radio interferometry imaging, analysis and simulation, relies on many lower-level components of the scientific Python ecosystem. In particular, the EHT collaboration used this library for the first imaging of a black hole. Within eht-imaging, NumPy arrays are used to store and manipulate numerical data at every step in the processing chain: from raw data through calibration and image reconstruction. SciPy supplies tools for general image-processing tasks such as filtering and image alignment, and scikit-image, an image-processing library that extends SciPy, provides higher-level functionality such as edge filters and Hough transforms. The ‘scipy.optimize’ module performs mathematical optimization. NetworkX[22](https://www.nature.com/articles/s41586-020-2649-2#ref-CR22 "Hagberg, A. A., Schult, D. A. & Swart, P. J. Exploring network structure, dynamics, and function using NetworkX. In Proc. 7th Python in Science Conf. \(eds Varoquaux, G., Vaught, T. & Millman, K. J.\) 11–15 \(2008\)."), a package for complex network analysis, is used to verify image comparison consistency. Astropy[23](https://www.nature.com/articles/s41586-020-2649-2#ref-CR23 "Astropy Collaboration et al. Astropy: a community Python package for astronomy. Astron. Astrophys. 558, A33 \(2013\)."),[24](https://www.nature.com/articles/s41586-020-2649-2#ref-CR24 "Price-Whelan, A. M. et al. The Astropy Project: building an open-science project and status of the v2.0 core package. Astron. J. 156, 123 \(2018\).") handles standard astronomical file formats and computes time–coordinate transformations. Matplotlib is used to visualize data and to generate the final image of the black hole.
The interactive environment created by the array programming foundation and the surrounding ecosystem of tools—inside of IPython or Jupyter—is ideally suited to exploratory data analysis. Users can fluidly inspect, manipulate and visualize their data, and rapidly iterate to refine programming statements. These statements are then stitched together into imperative or functional programs, or notebooks containing both computation and narrative. Scientific computing beyond exploratory work is often done in a text editor or an integrated development environment (IDE) such as Spyder. This rich and productive environment has made Python popular for scientific research.
To complement this facility for exploratory work and rapid prototyping, NumPy has developed a culture of using time-tested software engineering practices to improve collaboration and reduce error[30](https://www.nature.com/articles/s41586-020-2649-2#ref-CR30 "Millman, K. J. & Pérez, F. Developing open source scientific practice. In Implementing Reproducible Research \(eds Stodden, V., Leisch, F. & Peng, R. D.\) 149–183 \(CRC Press, 2014\). Describes the software engineering practices embraced by the NumPy and SciPy communities with a focus on how these practices improve research."). This culture is not only adopted by leaders in the project but also enthusiastically taught to newcomers. The NumPy team was early to adopt distributed revision control and code review to improve collaboration on code, and continuous testing that runs an extensive battery of automated tests for every proposed change to NumPy. The project also has comprehensive, high-quality documentation, integrated with the source code[31](https://www.nature.com/articles/s41586-020-2649-2#ref-CR31 "van der Walt, S. The SciPy Documentation Project \(technical overview\). In Proc. 7th Python in Science Conf. \(SciPy 2008\) \(eds Varoquaux, G., Vaught, T. & Millman, K. J.\) 27–28 \(2008\)."),[32](https://www.nature.com/articles/s41586-020-2649-2#ref-CR32 "Harrington, J. The SciPy Documentation Project. In Proc. 7th Python in Science Conference \(SciPy 2008\) \(eds Varoquaux, G., Vaught, T. & Millman, K. J.\) 33–35 \(2008\)."),[33](https://www.nature.com/articles/s41586-020-2649-2#ref-CR33 "Harrington, J. & Goldsmith, D. Progress report: NumPy and SciPy documentation in 2009. In Proc. 8th Python in Science Conf. \(SciPy 2009\) \(eds Varoquaux, G., van der Walt, S. & Millman, K. J.\) 84–87 \(2009\).").
This culture of using best practices for producing reliable scientific software has been adopted by the ecosystem of libraries that build on NumPy. For example, in a recent award given by the Royal Astronomical Society to Astropy, they state: “The Astropy Project has provided hundreds of junior scientists with experience in professional-standard software development practices including use of version control, unit testing, code review and issue tracking procedures. This is a vital skill set for modern researchers that is often missing from formal university education in physics or astronomy”[34](https://www.nature.com/articles/s41586-020-2649-2#ref-CR34 "Royal Astronomical Society Report of the RAS ‘A’ Awards Committee 2020: Astropy Project: 2020 Group Achievement Award \(A\) 
                  https://ras.ac.uk/sites/default/files/2020-01/Group%20Award%20-%20Astropy.pdf
                  
                 \(2020\)."). Community members explicitly work to address this lack of formal education through courses and workshops[35](https://www.nature.com/articles/s41586-020-2649-2#ref-CR35 "Wilson, G. Software carpentry: getting scientists to write better code by making them more productive. Comput. Sci. Eng. 8, 66–69 \(2006\)."),[36](https://www.nature.com/articles/s41586-020-2649-2#ref-CR36 "Hannay, J. E. et al. How do scientists develop and use scientific software? In Proc. 2009 ICSE Workshop on Software Engineering for Computational Science and Engineering 1–8 \(IEEE, 2009\)."),[37](https://www.nature.com/articles/s41586-020-2649-2#ref-CR37 "Millman, K. J., Brett, M., Barnowski, R. & Poline, J.-B. Teaching computational reproducibility for neuroimaging. Front. Neurosci. 12, 727 \(2018\).").
The recent rapid growth of data science, machine learning and artificial intelligence has further and dramatically boosted the scientific use of Python. Examples of its important applications, such as the eht-imaging library, now exist in almost every discipline in the natural and social sciences. These tools have become the primary software environment in many fields. NumPy and its ecosystem are commonly taught in university courses, boot camps and summer schools, and are the focus of community conferences and workshops worldwide. NumPy and its API have become truly ubiquitous.
## Array proliferation and interoperability
NumPy provides in-memory, multidimensional, homogeneously typed (that is, single-pointer and strided) arrays on CPUs. It runs on machines ranging from embedded devices to the world’s largest supercomputers, with performance approaching that of compiled languages. For most its existence, NumPy addressed the vast majority of array computation use cases.
However, scientific datasets now routinely exceed the memory capacity of a single machine and may be stored on multiple machines or in the cloud. In addition, the recent need to accelerate deep-learning and artificial intelligence applications has led to the emergence of specialized accelerator hardware, including graphics processing units (GPUs), tensor processing units (TPUs) and field-programmable gate arrays (FPGAs). Owing to its in-memory data model, NumPy is currently unable to directly utilize such storage and specialized hardware. However, both distributed data and also the parallel execution of GPUs, TPUs and FPGAs map well to the paradigm of array programming: therefore leading to a gap between available modern hardware architectures and the tools necessary to leverage their computational power.
The community’s efforts to fill this gap led to a proliferation of new array implementations. For example, each deep-learning framework created its own arrays; the PyTorch[38](https://www.nature.com/articles/s41586-020-2649-2#ref-CR38 "Paszke, A. et al. Pytorch: an imperative style, high-performance deep learning library. In Advances in Neural Information Processing Systems 32 \(eds Wallach, H. et al.\) 8024–8035 \(Neural Information Processing Systems, 2019\)."), Tensorflow[39](https://www.nature.com/articles/s41586-020-2649-2#ref-CR39 "Abadi, M. et al. TensorFlow: a system for large-scale machine learning. In OSDI’16: Proc. 12th USENIX Conf. Operating Systems Design and Implementation \(chairs Keeton, K. & Roscoe, T.\) 265–283 \(USENIX Association, 2016\)."), Apache MXNet[40](https://www.nature.com/articles/s41586-020-2649-2#ref-CR40 "Chen, T. et al. MXNet: a flexible and efficient machine learning library for heterogeneous distributed systems. Preprint at 
                  http://www.arxiv.org/abs/1512.01274
                  
                 \(2015\).") and JAX arrays all have the capability to run on CPUs and GPUs in a distributed fashion, using lazy evaluation to allow for additional performance optimizations. SciPy and PyData/Sparse both provide sparse arrays, which typically contain few non-zero values and store only those in memory for efficiency. In addition, there are projects that build on NumPy arrays as data containers, and extend its capabilities. Distributed arrays are made possible that way by Dask, and labelled arrays—referring to dimensions of an array by name rather than by index for clarity, compare x[:, 1] versus x.loc[:, 'time']—by xarray[41](https://www.nature.com/articles/s41586-020-2649-2#ref-CR41 "Hoyer, S. & Hamman, J. xarray: N–D labeled arrays and datasets in Python. J. Open Res. Softw. 5, 10 \(2017\).").
Such libraries often mimic the NumPy API, because this lowers the barrier to entry for newcomers and provides the wider community with a stable array programming interface. This, in turn, prevents disruptive schisms such as the divergence between Numeric and Numarray. But exploring new ways of working with arrays is experimental by nature and, in fact, several promising libraries (such as Theano and Caffe) have already ceased development. And each time that a user decides to try a new technology, they must change import statements and ensure that the new library implements all the parts of the NumPy API they currently use.
Ideally, operating on specialized arrays using NumPy functions or semantics would simply work, so that users could write code once, and would then benefit from switching between NumPy arrays, GPU arrays, distributed arrays and so forth as appropriate. To support array operations between external array objects, NumPy therefore added the capability to act as a central coordination mechanism with a well specified API (Fig. [2](https://www.nature.com/articles/s41586-020-2649-2#Fig2)).
To facilitate this interoperability, NumPy provides ‘protocols’ (or contracts of operation), that allow for specialized arrays to be passed to NumPy functions (Fig. [3](https://www.nature.com/articles/s41586-020-2649-2#Fig3)). NumPy, in turn, dispatches operations to the originating library, as required. Over four hundred of the most popular NumPy functions are supported. The protocols are implemented by widely used libraries such as Dask, CuPy, xarray and PyData/Sparse. Thanks to these developments, users can now, for example, scale their computation from a single machine to distributed systems using Dask. The protocols also compose well, allowing users to redeploy NumPy code at scale on distributed, multi-GPU systems via, for instance, CuPy arrays embedded in Dask arrays. Using NumPy’s high-level API, users can leverage highly parallel code execution on multiple systems with millions of cores, all with minimal code changes[42](https://www.nature.com/articles/s41586-020-2649-2#ref-CR42 "Entschev, P. Distributed multi-GPU computing with Dask, CuPy and RAPIDS. In EuroPython 2019 
                  https://ep2019.europython.eu/media/conference/slides/fX8dJsD-distributed-multi-gpu-computing-with-dask-cupy-and-rapids.pdf
                  
                 \(2019\).").
**Fig. 3: NumPy’s API and array protocols expose new arrays to the ecosystem.**
[![figure 3](https://media.springernature.com/lw685/springer-static/image/art%3A10.1038%2Fs41586-020-2649-2/MediaObjects/41586_2020_2649_Fig3_HTML.png)](https://www.nature.com/articles/s41586-020-2649-2/figures/3)
In this example, NumPy’s ‘mean’ function is called on a Dask array. The call succeeds by dispatching to the appropriate library implementation (in this case, Dask) and results in a new Dask array. Compare this code to the example code in Fig. [1g](https://www.nature.com/articles/s41586-020-2649-2#Fig1).
[Full size image](https://www.nature.com/articles/s41586-020-2649-2/figures/3)
These array protocols are now a key feature of NumPy, and are expected to only increase in importance. The NumPy developers—many of whom are authors of this Review—iteratively refine and add protocol designs to improve utility and simplify adoption.
## Discussion
NumPy combines the expressive power of array programming, the performance of C, and the readability, usability and versatility of Python in a mature, well tested, well documented and community-developed library. Libraries in the scientific Python ecosystem provide fast implementations of most important algorithms. Where extreme optimization is warranted, compiled languages can be used, such as Cython[43](https://www.nature.com/articles/s41586-020-2649-2#ref-CR43 "Behnel, S. et al. Cython: the best of both worlds. Comput. Sci. Eng. 13, 31–39 \(2011\)."), Numba[44](https://www.nature.com/articles/s41586-020-2649-2#ref-CR44 "Lam, S. K., Pitrou, A. & Seibert, S. Numba: a LLVM-based Python JIT compiler. In Proc. Second Workshop on the LLVM Compiler Infrastructure in HPC, LLVM ’15 7:1–7:6 \(ACM, 2015\).") and Pythran[45](https://www.nature.com/articles/s41586-020-2649-2#ref-CR45 "Guelton, S. et al. Pythran: enabling static optimization of scientific Python programs. Comput. Sci. Discov. 8, 014001 \(2015\)."); these languages extend Python and transparently accelerate bottlenecks. Owing to NumPy’s simple memory model, it is easy to write low-level, hand-optimized code, usually in C or Fortran, to manipulate NumPy arrays and pass them back to Python. Furthermore, using array protocols, it is possible to utilize the full spectrum of specialized hardware acceleration with minimal changes to existing code.
NumPy was initially developed by students, faculty and researchers to provide an advanced, open-source array programming library for Python, which was free to use and unencumbered by license servers and software protection dongles. There was a sense of building something consequential together for the benefit of many others. Participating in such an endeavour, within a welcoming community of like-minded individuals, held a powerful attraction for many early contributors.
These user–developers frequently had to write code from scratch to solve their own or their colleagues’ problems—often in low-level languages that preceded Python, such as Fortran[46](https://www.nature.com/articles/s41586-020-2649-2#ref-CR46 "Dongarra, J., Golub, G. H., Grosse, E., Moler, C. & Moore, K. Netlib and NA-Net: building a scientific computing community. IEEE Ann. Hist. Comput. 30, 30–41 \(2008\).") and C. To them, the advantages of an interactive, high-level array library were evident. The design of this new tool was informed by other powerful interactive programming languages for scientific computing such as Basis[47](https://www.nature.com/articles/s41586-020-2649-2#ref-CR47 "Barrett, K. A., Chiu, Y. H., Painter, J. F., Motteler, Z. C. & Dubois, P. F. Basis System, Part I: Running a Basis Program—A Tutorial for Beginners UCRL-MA-118543, Vol. 1 \(Lawrence Livermore National Laboratory 1995\)."),[48](https://www.nature.com/articles/s41586-020-2649-2#ref-CR48 "Dubois, P. F. & Motteler, Z. Basis System, Part II: Basis Language Reference Manual UCRL-MA-118543, Vol. 2 \(Lawrence Livermore National Laboratory, 1995\)."),[49](https://www.nature.com/articles/s41586-020-2649-2#ref-CR49 "Chiu, Y. H. & Dubois, P. F. Basis System, Part III: EZN User Manual UCRL-MA-118543, Vol. 3 \(Lawrence Livermore National Laboratory, 1995\)."),[50](https://www.nature.com/articles/s41586-020-2649-2#ref-CR50 "Chiu, Y. H. & Dubois, P. F. Basis System, Part IV: EZD User Manual UCRL-MA-118543, Vol. 4 \(Lawrence Livermore National Laboratory, 1995\)."), Yorick[51](https://www.nature.com/articles/s41586-020-2649-2#ref-CR51 "Munro, D. H. & Dubois, P. F. Using the Yorick interpreted language. Comput. Phys. 9, 609–615 \(1995\)."), R[52](https://www.nature.com/articles/s41586-020-2649-2#ref-CR52 "Ihaka, R. & Gentleman, R. R: a language for data analysis and graphics. J. Comput. Graph. Stat. 5, 299–314 \(1996\).") and APL[53](https://www.nature.com/articles/s41586-020-2649-2#ref-CR53 "Iverson, K. E. A programming language. In Proc. 1962 Spring Joint Computer Conf. 345–351 \(1962\)."), as well as commercial languages and environments such as IDL (Interactive Data Language) and MATLAB.
What began as an attempt to add an array object to Python became the foundation of a vibrant ecosystem of tools. Now, a large amount of scientific work depends on NumPy being correct, fast and stable. It is no longer a small community project, but core scientific infrastructure.
The developer culture has matured: although initial development was highly informal, NumPy now has a roadmap and a process for proposing and discussing large changes. The project has formal governance structures and is fiscally sponsored by NumFOCUS, a nonprofit that promotes open practices in research, data and scientific computing. Over the past few years, the project attracted its first funded development, sponsored by the Moore and Sloan Foundations, and received an award as part of the Chan Zuckerberg Initiative’s Essentials of Open Source Software programme. With this funding, the project was (and is) able to have sustained focus over multiple months to implement substantial new features and improvements. That said, the development of NumPy still depends heavily on contributions made by graduate students and researchers in their free time (see Supplementary Methods for more details).
NumPy is no longer merely the foundational array library underlying the scientific Python ecosystem, but it has become the standard API for tensor computation and a central coordinating mechanism between array types and technologies in Python. Work continues to expand on and improve these interoperability features.
Over the next decade, NumPy developers will face several challenges. New devices will be developed, and existing specialized hardware will evolve to meet diminishing returns on Moore’s law. There will be more, and a wider variety of, data science practitioners, a large proportion of whom will use NumPy. The scale of scientific data gathering will continue to increase, with the adoption of devices and instruments such as light-sheet microscopes and the Large Synoptic Survey Telescope (LSST)[54](https://www.nature.com/articles/s41586-020-2649-2#ref-CR54 "Jenness, T. et al. LSST data management software development practices and tools. In Proc. SPIE 10707, Software and Cyberinfrastructure for Astronomy V 1070709 \(SPIE and International Society for Optics and Photonics, 2018\)."). New generation languages, interpreters and compilers, such as Rust[55](https://www.nature.com/articles/s41586-020-2649-2#ref-CR55 "Matsakis, N. D. & Klock, F. S. The Rust language. Ada Letters 34, 103–104 \(2014\)."), Julia[56](https://www.nature.com/articles/s41586-020-2649-2#ref-CR56 "Bezanson, J., Edelman, A., Karpinski, S. & Shah, V. B. Julia: a fresh approach to numerical computing. SIAM Rev. 59, 65–98 \(2017\).") and LLVM[57](https://www.nature.com/articles/s41586-020-2649-2#ref-CR57 "Lattner, C. & Adve, V. LLVM: a compilation framework for lifelong program analysis and transformation. In Proc. 2004 Intl Symp. Code Generation and Optimization \(CGO’04\) 75–88 \(IEEE, 2004\)."), will create new concepts and data structures, and determine their viability.
Through the mechanisms described in this Review, NumPy is poised to embrace such a changing landscape, and to continue playing a leading part in interactive scientific computation, although to do so will require sustained funding from government, academia and industry. But, importantly, for NumPy to meet the needs of the next decade of data science, it will also need a new generation of graduate students and community contributors to drive it forward.
## References
  1. Abbott, B. P. et al. Observation of gravitational waves from a binary black hole merger. _Phys. Rev. Lett_. **116** , 061102 (2016).
[Article](https://doi.org/10.1103%2FPhysRevLett.116.061102) [ADS](http://adsabs.harvard.edu/cgi-bin/nph-data_query?link_type=ABSTRACT&bibcode=2016PhRvL.116f1102A) [MathSciNet](http://www.ams.org/mathscinet-getitem?mr=3707758) [CAS](https://www.nature.com/articles/cas-redirect/1:STN:280:DC%2BC28jltVGjug%3D%3D) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Observation%20of%20gravitational%20waves%20from%20a%20binary%20black%20hole%20merger&journal=Phys.%20Rev.%20Lett.&doi=10.1103%2FPhysRevLett.116.061102&volume=116&publication_year=2016&author=Abbott%2CBP)
  2. Chael, A. et al. High-resolution linear polarimetric imaging for the Event Horizon Telescope. _Astrophys. J_. **286** , 11 (2016).
[Article](https://doi.org/10.3847%2F0004-637X%2F829%2F1%2F11) [ADS](http://adsabs.harvard.edu/cgi-bin/nph-data_query?link_type=ABSTRACT&bibcode=2016ApJ...833...11C) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=High-resolution%20linear%20polarimetric%20imaging%20for%20the%20Event%20Horizon%20Telescope&journal=Astrophys.%20J.&doi=10.3847%2F0004-637X%2F829%2F1%2F11&volume=286&publication_year=2016&author=Chael%2CA)
  3. Dubois, P. F., Hinsen, K. & Hugunin, J. Numerical Python. _Comput. Phys_. **10** , 262–267 (1996).
[Article](https://doi.org/10.1063%2F1.4822400) [ADS](http://adsabs.harvard.edu/cgi-bin/nph-data_query?link_type=ABSTRACT&bibcode=1996ComPh..10..262D) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Numerical%20Python&journal=Comput.%20Phys.&doi=10.1063%2F1.4822400&volume=10&pages=262-267&publication_year=1996&author=Dubois%2CPF&author=Hinsen%2CK&author=Hugunin%2CJ)
  4. Ascher, D., Dubois, P. F., Hinsen, K., Hugunin, J. & Oliphant, T. E. _An Open Source Project: Numerical Python_ (Lawrence Livermore National Laboratory, 2001).
  5. Yang, T.-Y., Furnish, G. & Dubois, P. F. Steering object-oriented scientific computations. In _Proc. TOOLS USA 97. Intl Conf. Technology of Object Oriented Systems and Languages_ (eds Ege, R., Singh, M. & Meyer, B.) 112–119 (IEEE, 1997).
  6. Greenfield, P., Miller, J. T., Hsu, J. & White, R. L. numarray: a new scientific array package for Python. In _PyCon DC 2003_ <http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.112.9899> (2003).
  7. Oliphant, T. E. _Guide to NumPy_ 1st edn (Trelgol Publishing, 2006).
  8. Dubois, P. F. Python: batteries included. _Comput. Sci. Eng_. **9** , 7–9 (2007).
[Article](https://doi.org/10.1109%2FMCSE.2007.51) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Python%3A%20batteries%20included&journal=Comput.%20Sci.%20Eng.&doi=10.1109%2FMCSE.2007.51&volume=9&pages=7-9&publication_year=2007&author=Dubois%2CPF)
  9. Oliphant, T. E. Python for scientific computing. _Comput. Sci. Eng_. **9** , 10–20 (2007).
[Article](https://doi.org/10.1109%2FMCSE.2007.58) [CAS](https://www.nature.com/articles/cas-redirect/1:CAS:528:DC%2BD2sXltVWiurw%3D) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Python%20for%20scientific%20computing&journal=Comput.%20Sci.%20Eng.&doi=10.1109%2FMCSE.2007.58&volume=9&pages=10-20&publication_year=2007&author=Oliphant%2CTE)
  10. Millman, K. J. & Aivazis, M. Python for scientists and engineers. _Comput. Sci. Eng_. **13** , 9–12 (2011).
[Article](https://doi.org/10.1109%2FMCSE.2011.36) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Python%20for%20scientists%20and%20engineers&journal=Comput.%20Sci.%20Eng.&doi=10.1109%2FMCSE.2011.36&volume=13&pages=9-12&publication_year=2011&author=Millman%2CKJ&author=Aivazis%2CM)
  11. Pérez, F., Granger, B. E. & Hunter, J. D. Python: an ecosystem for scientific computing. _Comput. Sci. Eng_. **13** , 13–21 (2011). **Explains why the scientific Python ecosystem is a highly productive environment for research**.
[Article](https://doi.org/10.1109%2FMCSE.2010.119) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Python%3A%20an%20ecosystem%20for%20scientific%20computing&journal=Comput.%20Sci.%20Eng.&doi=10.1109%2FMCSE.2010.119&volume=13&pages=13-21&publication_year=2011&author=P%C3%A9rez%2CF&author=Granger%2CBE&author=Hunter%2CJD)
  12. Virtanen, P. et al. SciPy 1.0—fundamental algorithms for scientific computing in Python. _Nat. Methods_ **17** , 261–272 (2020); correction **17** , 352 (2020). **Introduces the SciPy library and includes a more detailed history of NumPy and SciPy.**
[Article](https://doi.org/10.1038%2Fs41592-019-0686-2) [CAS](https://www.nature.com/articles/cas-redirect/1:CAS:528:DC%2BB3cXislCjuro%3D) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=SciPy%201.0%E2%80%94fundamental%20algorithms%20for%20scientific%20computing%20in%20Python&journal=Nat.%20Methods&doi=10.1038%2Fs41592-019-0686-2&volume=17&pages=261-272&publication_year=2020&author=Virtanen%2CP)
  13. Hunter, J. D. Matplotlib: a 2D graphics environment. _Comput. Sci. Eng_. **9** , 90–95 (2007).
[Article](https://doi.org/10.1109%2FMCSE.2007.55) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Matplotlib%3A%20a%202D%20graphics%20environment&journal=Comput.%20Sci.%20Eng.&doi=10.1109%2FMCSE.2007.55&volume=9&pages=90-95&publication_year=2007&author=Hunter%2CJD)
  14. McKinney, W. Data structures for statistical computing in Python. In _Proc. 9th Python in Science Conf_. (eds van der Walt, S. & Millman, K. J.) 56–61 (2010).
  15. Pedregosa, F. et al. Scikit-learn: machine learning in Python. _J. Mach. Learn. Res_. **12** , 2825–2830 (2011).
[MathSciNet](http://www.ams.org/mathscinet-getitem?mr=2854348) [MATH](http://www.emis.de/MATH-item?1280.68189) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Scikit-learn%3A%20machine%20learning%20in%20Python&journal=J.%20Mach.%20Learn.%20Res.&volume=12&pages=2825-2830&publication_year=2011&author=Pedregosa%2CF)
  16. van der Walt, S. et al. scikit-image: image processing in Python. _PeerJ_ **2** , e453 (2014).
[Article](https://doi.org/10.7717%2Fpeerj.453) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=scikit-image%3A%20image%20processing%20in%20Python&journal=PeerJ&doi=10.7717%2Fpeerj.453&volume=2&publication_year=2014&author=Walt%2CS)
  17. van der Walt, S., Colbert, S. C. & Varoquaux, G. The NumPy array: a structure for efficient numerical computation. _Comput. Sci. Eng_. **13** , 22–30 (2011). **Discusses the NumPy array data structure with a focus on how it enables efficient computation**.
  18. Wang, Q., Zhang, X., Zhang, Y. & Yi, Q. AUGEM: automatically generate high performance dense linear algebra kernels on x86 CPUs. In _SC’13: Proc. Intl Conf. High Performance Computing, Networking, Storage and Analysis_ 25 (IEEE, 2013).
  19. Xianyi, Z., Qian, W. & Yunquan, Z. Model-driven level 3 BLAS performance optimization on Loongson 3A processor. In _2012 IEEE 18th Intl Conf. Parallel and Distributed Systems_ 684–691 (IEEE, 2012).
  20. Pérez, F. & Granger, B. E. IPython: a system for interactive scientific computing. _Comput. Sci. Eng_. **9** , 21–29 (2007).
[Article](https://doi.org/10.1109%2FMCSE.2007.53) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=IPython%3A%20a%20system%20for%20interactive%20scientific%20computing&journal=Comput.%20Sci.%20Eng.&doi=10.1109%2FMCSE.2007.53&volume=9&pages=21-29&publication_year=2007&author=P%C3%A9rez%2CF&author=Granger%2CBE)
  21. Kluyver, T. et al. Jupyter Notebooks—a publishing format for reproducible computational workflows. In _Positioning and Power in Academic Publishing: Players, Agents and Agendas_ (eds Loizides, F. & Schmidt, B.) 87–90 (IOS Press, 2016).
  22. Hagberg, A. A., Schult, D. A. & Swart, P. J. Exploring network structure, dynamics, and function using NetworkX. In _Proc. 7th Python in Science Conf_. (eds Varoquaux, G., Vaught, T. & Millman, K. J.) 11–15 (2008).
  23. Astropy Collaboration et al. Astropy: a community Python package for astronomy. _Astron. Astrophys_. **558** , A33 (2013).
[Article](https://doi.org/10.1051%2F0004-6361%2F201322068) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Astropy%3A%20a%20community%20Python%20package%20for%20astronomy&journal=Astron.%20Astrophys.&doi=10.1051%2F0004-6361%2F201322068&volume=558&publication_year=2013&author=Astropy%20Collaboration%2C)
  24. Price-Whelan, A. M. et al. The Astropy Project: building an open-science project and status of the v2.0 core package. _Astron. J_. **156** , 123 (2018).
[Article](https://doi.org/10.3847%2F1538-3881%2Faac387) [ADS](http://adsabs.harvard.edu/cgi-bin/nph-data_query?link_type=ABSTRACT&bibcode=2018AJ....156..123A) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=The%20Astropy%20Project%3A%20building%20an%20open-science%20project%20and%20status%20of%20the%20v2.0%20core%20package&journal=Astron.%20J.&doi=10.3847%2F1538-3881%2Faac387&volume=156&publication_year=2018&author=Price-Whelan%2CAM)
  25. Cock, P. J. et al. Biopython: freely available Python tools for computational molecular biology and bioinformatics. _Bioinformatics_ **25** , 1422–1423 (2009).
[Article](https://doi.org/10.1093%2Fbioinformatics%2Fbtp163) [CAS](https://www.nature.com/articles/cas-redirect/1:CAS:528:DC%2BD1MXmtFeqt74%3D) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Biopython%3A%20freely%20available%20Python%20tools%20for%20computational%20molecular%20biology%20and%20bioinformatics&journal=Bioinformatics&doi=10.1093%2Fbioinformatics%2Fbtp163&volume=25&pages=1422-1423&publication_year=2009&author=Cock%2CPJ)
  26. Millman, K. J. & Brett, M. Analysis of functional magnetic resonance imaging in Python. _Comput. Sci. Eng_. **9** , 52–55 (2007).
[Article](https://doi.org/10.1109%2FMCSE.2007.46) [CAS](https://www.nature.com/articles/cas-redirect/1:CAS:528:DC%2BD2sXltVWiurs%3D) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Analysis%20of%20functional%20magnetic%20resonance%20imaging%20in%20Python&journal=Comput.%20Sci.%20Eng.&doi=10.1109%2FMCSE.2007.46&volume=9&pages=52-55&publication_year=2007&author=Millman%2CKJ&author=Brett%2CM)
  27. The SunPy Community et al. SunPy—Python for solar physics. _Comput. Sci. Discov_. **8** , 014009 (2015).
[Article](https://doi.org/10.1088%2F1749-4699%2F8%2F1%2F014009) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=SunPy%E2%80%94Python%20for%20solar%20physics&journal=Comput.%20Sci.%20Discov.&doi=10.1088%2F1749-4699%2F8%2F1%2F014009&volume=8&publication_year=2015)
  28. Hamman, J., Rocklin, M. & Abernathy, R. Pangeo: a big-data ecosystem for scalable Earth system science. In _EGU General Assembly Conf. Abstracts_ 12146 (2018).
  29. Chael, A. A. et al. ehtim: imaging, analysis, and simulation software for radio interferometry. _Astrophysics Source Code Library_ <https://ascl.net/1904.004> (2019).
  30. Millman, K. J. & Pérez, F. Developing open source scientific practice. In _Implementing Reproducible Research_ (eds Stodden, V., Leisch, F. & Peng, R. D.) 149–183 (CRC Press, 2014). **Describes the software engineering practices embraced by the NumPy and SciPy communities with a focus on how these practices improve research**.
  31. van der Walt, S. The SciPy Documentation Project (technical overview). In _Proc. 7th Python in Science Conf. (SciPy 2008)_ (eds Varoquaux, G., Vaught, T. & Millman, K. J.) 27–28 (2008).
  32. Harrington, J. The SciPy Documentation Project. In _Proc. 7th Python in Science Conference (SciPy 2008)_ (eds Varoquaux, G., Vaught, T. & Millman, K. J.) 33–35 (2008).
  33. Harrington, J. & Goldsmith, D. Progress report: NumPy and SciPy documentation in 2009. In _Proc. 8th Python in Science Conf. (SciPy 2009)_ (eds Varoquaux, G., van der Walt, S. & Millman, K. J.) 84–87 (2009).
  34. Royal Astronomical Society _Report of the RAS ‘A’ Awards Committee 2020: Astropy Project: 2020 Group Achievement Award (A)_ <https://ras.ac.uk/sites/default/files/2020-01/Group%20Award%20-%20Astropy.pdf> (2020).
  35. Wilson, G. Software carpentry: getting scientists to write better code by making them more productive. _Comput. Sci. Eng_. **8** , 66–69 (2006).
[Article](https://doi.org/10.1109%2FMCSE.2006.122) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Software%20carpentry%3A%20getting%20scientists%20to%20write%20better%20code%20by%20making%20them%20more%20productive&journal=Comput.%20Sci.%20Eng.&doi=10.1109%2FMCSE.2006.122&volume=8&pages=66-69&publication_year=2006&author=Wilson%2CG)
  36. Hannay, J. E. et al. How do scientists develop and use scientific software? In _Proc. 2009 ICSE Workshop on Software Engineering for Computational Science and Engineering_ 1–8 (IEEE, 2009).
  37. Millman, K. J., Brett, M., Barnowski, R. & Poline, J.-B. Teaching computational reproducibility for neuroimaging. _Front. Neurosci_. **12** , 727 (2018).
[Article](https://doi.org/10.3389%2Ffnins.2018.00727) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Teaching%20computational%20reproducibility%20for%20neuroimaging&journal=Front.%20Neurosci.&doi=10.3389%2Ffnins.2018.00727&volume=12&publication_year=2018&author=Millman%2CKJ&author=Brett%2CM&author=Barnowski%2CR&author=Poline%2CJ-B)
  38. Paszke, A. et al. Pytorch: an imperative style, high-performance deep learning library. In _Advances in Neural Information Processing Systems 32_ (eds Wallach, H. et al.) 8024–8035 (Neural Information Processing Systems, 2019).
  39. Abadi, M. et al. TensorFlow: a system for large-scale machine learning. In _OSDI’16: Proc. 12th USENIX Conf. Operating Systems Design and Implementation_ (chairs Keeton, K. & Roscoe, T.) 265–283 (USENIX Association, 2016).
  40. Chen, T. et al. MXNet: a flexible and efficient machine learning library for heterogeneous distributed systems. Preprint at <http://www.arxiv.org/abs/1512.01274> (2015).
  41. Hoyer, S. & Hamman, J. xarray: N–D labeled arrays and datasets in Python. _J. Open Res. Softw_. **5** , 10 (2017).
[Article](https://doi.org/10.5334%2Fjors.148) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=xarray%3A%20N%E2%80%93D%20labeled%20arrays%20and%20datasets%20in%20Python&journal=J.%20Open%20Res.%20Softw.&doi=10.5334%2Fjors.148&volume=5&publication_year=2017&author=Hoyer%2CS&author=Hamman%2CJ)
  42. Entschev, P. Distributed multi-GPU computing with Dask, CuPy and RAPIDS. In _EuroPython 2019_ <https://ep2019.europython.eu/media/conference/slides/fX8dJsD-distributed-multi-gpu-computing-with-dask-cupy-and-rapids.pdf> (2019).
  43. Behnel, S. et al. Cython: the best of both worlds. _Comput. Sci. Eng_. **13** , 31–39 (2011).
[Article](https://doi.org/10.1109%2FMCSE.2010.118) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Cython%3A%20the%20best%20of%20both%20worlds&journal=Comput.%20Sci.%20Eng.&doi=10.1109%2FMCSE.2010.118&volume=13&pages=31-39&publication_year=2011&author=Behnel%2CS)
  44. Lam, S. K., Pitrou, A. & Seibert, S. Numba: a LLVM-based Python JIT compiler. In _Proc. Second Workshop on the LLVM Compiler Infrastructure in HPC, LLVM ’15_ 7:1–7:6 (ACM, 2015).
  45. Guelton, S. et al. Pythran: enabling static optimization of scientific Python programs. _Comput. Sci. Discov_. **8** , 014001 (2015).
[Article](https://doi.org/10.1088%2F1749-4680%2F8%2F1%2F014001) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Pythran%3A%20enabling%20static%20optimization%20of%20scientific%20Python%20programs&journal=Comput.%20Sci.%20Discov.&doi=10.1088%2F1749-4680%2F8%2F1%2F014001&volume=8&publication_year=2015&author=Guelton%2CS)
  46. Dongarra, J., Golub, G. H., Grosse, E., Moler, C. & Moore, K. Netlib and NA-Net: building a scientific computing community. _IEEE Ann. Hist. Comput_. **30** , 30–41 (2008).
[Article](https://doi.org/10.1109%2FMAHC.2008.29) [MathSciNet](http://www.ams.org/mathscinet-getitem?mr=2432392) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Netlib%20and%20NA-Net%3A%20building%20a%20scientific%20computing%20community&journal=IEEE%20Ann.%20Hist.%20Comput.&doi=10.1109%2FMAHC.2008.29&volume=30&pages=30-41&publication_year=2008&author=Dongarra%2CJ&author=Golub%2CGH&author=Grosse%2CE&author=Moler%2CC&author=Moore%2CK)
  47. Barrett, K. A., Chiu, Y. H., Painter, J. F., Motteler, Z. C. & Dubois, P. F. _Basis System, Part I: Running a Basis Program—A Tutorial for Beginners_ UCRL-MA-118543, Vol. 1 (Lawrence Livermore National Laboratory 1995).
  48. Dubois, P. F. & Motteler, Z. _Basis System, Part II: Basis Language Reference Manual_ UCRL-MA-118543, Vol. 2 (Lawrence Livermore National Laboratory, 1995).
  49. Chiu, Y. H. & Dubois, P. F. _Basis System, Part III: EZN User Manual_ UCRL-MA-118543, Vol. 3 (Lawrence Livermore National Laboratory, 1995).
  50. Chiu, Y. H. & Dubois, P. F. _Basis System, Part IV: EZD User Manual_ UCRL-MA-118543, Vol. 4 (Lawrence Livermore National Laboratory, 1995).
  51. Munro, D. H. & Dubois, P. F. Using the Yorick interpreted language. _Comput. Phys_. **9** , 609–615 (1995).
[Article](https://doi.org/10.1063%2F1.4823451) [ADS](http://adsabs.harvard.edu/cgi-bin/nph-data_query?link_type=ABSTRACT&bibcode=1995ComPh...9..609M) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Using%20the%20Yorick%20interpreted%20language&journal=Comput.%20Phys.&doi=10.1063%2F1.4823451&volume=9&pages=609-615&publication_year=1995&author=Munro%2CDH&author=Dubois%2CPF)
  52. Ihaka, R. & Gentleman, R. R: a language for data analysis and graphics. _J. Comput. Graph. Stat_. **5** , 299–314 (1996).
[ Google Scholar](http://scholar.google.com/scholar_lookup?&title=R%3A%20a%20language%20for%20data%20analysis%20and%20graphics&journal=J.%20Comput.%20Graph.%20Stat.&volume=5&pages=299-314&publication_year=1996&author=Ihaka%2CR&author=Gentleman%2CR)
  53. Iverson, K. E. A programming language. In _Proc. 1962 Spring Joint Computer Conf_. 345–351 (1962).
  54. Jenness, T. et al. LSST data management software development practices and tools. In _Proc. SPIE 10707, Software and Cyberinfrastructure for Astronomy V_ 1070709 (SPIE and International Society for Optics and Photonics, 2018).
  55. Matsakis, N. D. & Klock, F. S. The Rust language. _Ada Letters_ **34** , 103–104 (2014).
[Article](https://doi.org/10.1145%2F2692956.2663188) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=The%20Rust%20language&journal=Ada%20Letters&doi=10.1145%2F2692956.2663188&volume=34&pages=103-104&publication_year=2014&author=Matsakis%2CND&author=Klock%2CFS)
  56. Bezanson, J., Edelman, A., Karpinski, S. & Shah, V. B. Julia: a fresh approach to numerical computing. _SIAM Rev_. **59** , 65–98 (2017).
[Article](https://doi.org/10.1137%2F141000671) [MathSciNet](http://www.ams.org/mathscinet-getitem?mr=3605826) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Julia%3A%20a%20fresh%20approach%20to%20numerical%20computing&journal=SIAM%20Rev.&doi=10.1137%2F141000671&volume=59&pages=65-98&publication_year=2017&author=Bezanson%2CJ&author=Edelman%2CA&author=Karpinski%2CS&author=Shah%2CVB)
  57. Lattner, C. & Adve, V. LLVM: a compilation framework for lifelong program analysis and transformation. In _Proc. 2004 Intl Symp. Code Generation and Optimization (CGO’04)_ 75–88 (IEEE, 2004).


[Download references](https://citation-needed.springer.com/v2/references/10.1038/s41586-020-2649-2?format=refman&flavour=references)
## Acknowledgements
We thank R. Barnowski, P. Dubois, M. Eickenberg, and P. Greenfield, who suggested text and provided helpful feedback on the manuscript. K.J.M. and S.J.v.d.W. were funded in part by the Gordon and Betty Moore Foundation through grant GBMF3834 and by the Alfred P. Sloan Foundation through grant 2013-10-27 to the University of California, Berkeley. S.J.v.d.W., S.B., M.P. and W.W. were funded in part by the Gordon and Betty Moore Foundation through grant GBMF5447 and by the Alfred P. Sloan Foundation through grant G-2017-9960 to the University of California, Berkeley.
## Author information
### Authors and Affiliations
  1. Independent researcher, Logan, UT, USA
Charles R. Harris
  2. Brain Imaging Center, University of California, Berkeley, Berkeley, CA, USA
K. Jarrod Millman, Stéfan J. van der Walt & Matthew Brett
  3. Division of Biostatistics, University of California, Berkeley, Berkeley, CA, USA
K. Jarrod Millman
  4. Berkeley Institute for Data Science, University of California, Berkeley, Berkeley, CA, USA
K. Jarrod Millman, Stéfan J. van der Walt, Sebastian Berg, Matti Picus & Warren Weckesser
  5. Applied Mathematics, Stellenbosch University, Stellenbosch, South Africa
Stéfan J. van der Walt
  6. Quansight, Austin, TX, USA
Ralf Gommers, Pearu Peterson, Hameer Abbasi & Travis E. Oliphant
  7. Department of Physics, University of Jyväskylä, Jyväskylä, Finland
Pauli Virtanen
  8. Nanoscience Center, University of Jyväskylä, Jyväskylä, Finland
Pauli Virtanen
  9. Mercari JP, Tokyo, Japan
David Cournapeau
  10. Department of Engineering, University of Cambridge, Cambridge, UK
Eric Wieser
  11. Independent researcher, Karlsruhe, Germany
Julian Taylor
  12. Independent researcher, Berkeley, CA, USA
Nathaniel J. Smith
  13. Enthought, Austin, TX, USA
Robert Kern
  14. Google Research, Mountain View, CA, USA
Stephan Hoyer
  15. Department of Astronomy and Astrophysics, University of Toronto, Toronto, Ontario, Canada
Marten H. van Kerkwijk
  16. School of Psychology, University of Birmingham, Edgbaston, Birmingham, UK
Matthew Brett
  17. Department of Physics, Temple University, Philadelphia, PA, USA
Allan Haldane
  18. Google, Zurich, Switzerland
Jaime Fernández del Río
  19. Department of Physics and Astronomy, The University of British Columbia, Vancouver, British Columbia, Canada
Mark Wiebe
  20. Amazon, Seattle, WA, USA
Mark Wiebe
  21. Independent researcher, Saue, Estonia
Pearu Peterson
  22. Department of Mechanics and Applied Mathematics, Institute of Cybernetics at Tallinn Technical University, Tallinn, Estonia
Pearu Peterson
  23. Department of Biological and Agricultural Engineering, University of Georgia, Athens, GA, USA
Pierre Gérard-Marchant
  24. France-IX Services, Paris, France
Pierre Gérard-Marchant
  25. Department of Economics, University of Oxford, Oxford, UK
Kevin Sheppard
  26. CCS-7, Los Alamos National Laboratory, Los Alamos, NM, USA
Tyler Reddy
  27. Laboratory for Fluorescence Dynamics, Biomedical Engineering Department, University of California, Irvine, Irvine, CA, USA
Christoph Gohlke


Authors
  1. Charles R. Harris
[View author publications](https://www.nature.com/search?author=Charles%20R.%20Harris)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Charles%20R.%20Harris)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Charles%20R.%20Harris%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  2. K. Jarrod Millman
[View author publications](https://www.nature.com/search?author=K.%20Jarrod%20Millman)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=K.%20Jarrod%20Millman)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22K.%20Jarrod%20Millman%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  3. Stéfan J. van der Walt
[View author publications](https://www.nature.com/search?author=St%C3%A9fan%20J.%20van%20der%20Walt)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=St%C3%A9fan%20J.%20van%20der%20Walt)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22St%C3%A9fan%20J.%20van%20der%20Walt%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  4. Ralf Gommers
[View author publications](https://www.nature.com/search?author=Ralf%20Gommers)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Ralf%20Gommers)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Ralf%20Gommers%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  5. Pauli Virtanen
[View author publications](https://www.nature.com/search?author=Pauli%20Virtanen)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Pauli%20Virtanen)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Pauli%20Virtanen%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  6. David Cournapeau
[View author publications](https://www.nature.com/search?author=David%20Cournapeau)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=David%20Cournapeau)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22David%20Cournapeau%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  7. Eric Wieser
[View author publications](https://www.nature.com/search?author=Eric%20Wieser)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Eric%20Wieser)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Eric%20Wieser%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  8. Julian Taylor
[View author publications](https://www.nature.com/search?author=Julian%20Taylor)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Julian%20Taylor)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Julian%20Taylor%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  9. Sebastian Berg
[View author publications](https://www.nature.com/search?author=Sebastian%20Berg)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Sebastian%20Berg)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Sebastian%20Berg%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  10. Nathaniel J. Smith
[View author publications](https://www.nature.com/search?author=Nathaniel%20J.%20Smith)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Nathaniel%20J.%20Smith)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Nathaniel%20J.%20Smith%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  11. Robert Kern
[View author publications](https://www.nature.com/search?author=Robert%20Kern)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Robert%20Kern)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Robert%20Kern%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  12. Matti Picus
[View author publications](https://www.nature.com/search?author=Matti%20Picus)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Matti%20Picus)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Matti%20Picus%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  13. Stephan Hoyer
[View author publications](https://www.nature.com/search?author=Stephan%20Hoyer)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Stephan%20Hoyer)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Stephan%20Hoyer%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  14. Marten H. van Kerkwijk
[View author publications](https://www.nature.com/search?author=Marten%20H.%20van%20Kerkwijk)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Marten%20H.%20van%20Kerkwijk)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Marten%20H.%20van%20Kerkwijk%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  15. Matthew Brett
[View author publications](https://www.nature.com/search?author=Matthew%20Brett)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Matthew%20Brett)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Matthew%20Brett%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  16. Allan Haldane
[View author publications](https://www.nature.com/search?author=Allan%20Haldane)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Allan%20Haldane)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Allan%20Haldane%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  17. Jaime Fernández del Río
[View author publications](https://www.nature.com/search?author=Jaime%20Fern%C3%A1ndez%20del%20R%C3%ADo)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Jaime%20Fern%C3%A1ndez%20del%20R%C3%ADo)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Jaime%20Fern%C3%A1ndez%20del%20R%C3%ADo%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  18. Mark Wiebe
[View author publications](https://www.nature.com/search?author=Mark%20Wiebe)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Mark%20Wiebe)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Mark%20Wiebe%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  19. Pearu Peterson
[View author publications](https://www.nature.com/search?author=Pearu%20Peterson)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Pearu%20Peterson)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Pearu%20Peterson%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  20. Pierre Gérard-Marchant
[View author publications](https://www.nature.com/search?author=Pierre%20G%C3%A9rard-Marchant)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Pierre%20G%C3%A9rard-Marchant)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Pierre%20G%C3%A9rard-Marchant%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  21. Kevin Sheppard
[View author publications](https://www.nature.com/search?author=Kevin%20Sheppard)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Kevin%20Sheppard)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Kevin%20Sheppard%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  22. Tyler Reddy
[View author publications](https://www.nature.com/search?author=Tyler%20Reddy)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Tyler%20Reddy)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Tyler%20Reddy%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  23. Warren Weckesser
[View author publications](https://www.nature.com/search?author=Warren%20Weckesser)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Warren%20Weckesser)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Warren%20Weckesser%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  24. Hameer Abbasi
[View author publications](https://www.nature.com/search?author=Hameer%20Abbasi)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Hameer%20Abbasi)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Hameer%20Abbasi%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  25. Christoph Gohlke
[View author publications](https://www.nature.com/search?author=Christoph%20Gohlke)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Christoph%20Gohlke)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Christoph%20Gohlke%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)
  26. Travis E. Oliphant
[View author publications](https://www.nature.com/search?author=Travis%20E.%20Oliphant)
Search author on:[PubMed](https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=search&term=Travis%20E.%20Oliphant)[Google Scholar](https://scholar.google.co.uk/scholar?as_q=&num=10&btnG=Search+Scholar&as_epq=&as_oq=&as_eq=&as_occt=any&as_sauthors=%22Travis%20E.%20Oliphant%22&as_publication=&as_ylo=&as_yhi=&as_allsubj=all&hl=en)


### Contributions
K.J.M. and S.J.v.d.W. composed the manuscript with input from others. S.B., R.G., K.S., W.W., M.B. and T.R. contributed text. All authors contributed substantial code, documentation and/or expertise to the NumPy project. All authors reviewed the manuscript.
### Corresponding authors
Correspondence to K. Jarrod Millman, Stéfan J. van der Walt or Ralf Gommers.
## Ethics declarations
### Competing interests
The authors declare no competing interests.
## Additional information
**Peer review information** _Nature_ thanks Edouard Duchesnay, Alan Edelman and the other, anonymous, reviewer(s) for their contribution to the peer review of this work.
**Publisher’s note** Springer Nature remains neutral with regard to jurisdictional claims in published maps and institutional affiliations.
## Supplementary information
### [Supplementary Information](https://static-content.springer.com/esm/art%3A10.1038%2Fs41586-020-2649-2/MediaObjects/41586_2020_2649_MOESM1_ESM.pdf)
This file contains Supplementary Methods, including Supplementary Figure 1 and additional references.
## Rights and permissions
**Open Access** This article is licensed under a Creative Commons Attribution 4.0 International License, which permits use, sharing, adaptation, distribution and reproduction in any medium or format, as long as you give appropriate credit to the original author(s) and the source, provide a link to the Creative Commons license, and indicate if changes were made. The images or other third party material in this article are included in the article’s Creative Commons license, unless indicated otherwise in a credit line to the material. If material is not included in the article’s Creative Commons license and your intended use is not permitted by statutory regulation or exceeds the permitted use, you will need to obtain permission directly from the copyright holder. To view a copy of this license, visit <http://creativecommons.org/licenses/by/4.0/>.
[Reprints and permissions](https://s100.copyright.com/AppDispatchServlet?title=Array%20programming%20with%20NumPy&author=Charles%20R.%20Harris%20et%20al&contentID=10.1038%2Fs41586-020-2649-2&copyright=The%20Author%28s%29&publication=0028-0836&publicationDate=2020-09-16&publisherName=SpringerNature&orderBeanReset=true&oa=CC%20BY)
## About this article
[![Check for updates. Verify currency and authenticity via CrossMark](https://www.nature.com/articles/s41586-020-2649-2)](https://crossmark.crossref.org/dialog/?doi=10.1038/s41586-020-2649-2)
### Cite this article
Harris, C.R., Millman, K.J., van der Walt, S.J. _et al._ Array programming with NumPy. _Nature_ **585** , 357–362 (2020). https://doi.org/10.1038/s41586-020-2649-2
[Download citation](https://citation-needed.springer.com/v2/references/10.1038/s41586-020-2649-2?format=refman&flavour=citation)
  * Received: 21 February 2020
  * Accepted: 17 June 2020
  * Published: 16 September 2020
  * Version of record: 16 September 2020
  * Issue date: 17 September 2020
  * DOI: https://doi.org/10.1038/s41586-020-2649-2


### Share this article
Anyone you share the following link with will be able to read this content:
Get shareable link
Sorry, a shareable link is not currently available for this article.
Copy shareable link to clipboard
Provided by the Springer Nature SharedIt content-sharing initiative 
### Subjects
  * [Computational neuroscience](https://www.nature.com/subjects/computational-neuroscience)
  * [Computational science](https://www.nature.com/subjects/computational-science)
  * [Computer science](https://www.nature.com/subjects/computer-science)
  * [Software](https://www.nature.com/subjects/software)
  * [Solar physics](https://www.nature.com/subjects/solar-physics)


## This article is cited by
  * ###  [ Bi-temporal change detection for topographic map updates using panoptic segmentation of VNIR orthophotos and LiDAR data ](https://doi.org/10.1007/s10707-025-00561-z)
    * Maciej Adamiak
    * Krzysztof Będkowski
    * Szymon Wójcik
_GeoInformatica_ (2026)
  * ###  [ Descriptive epidemiology and surgical outcomes of intracranial tumors in southwestern uganda: a cross-sectional study ](https://doi.org/10.1007/s11060-025-05249-w)
    * David Kitya
    * Joseph Mary Ssembatya
    * Alvan-Emeka K. Ukachukwu
_Journal of Neuro-Oncology_ (2026)


[ Download PDF ](https://www.nature.com/articles/s41586-020-2649-2.pdf)
  * Sections
  * Figures
  * References


  * [Abstract](https://www.nature.com/articles/s41586-020-2649-2#Abs1)
  * [Main](https://www.nature.com/articles/s41586-020-2649-2#Sec1)
  * [NumPy arrays](https://www.nature.com/articles/s41586-020-2649-2#Sec2)
  * [Scientific Python ecosystem](https://www.nature.com/articles/s41586-020-2649-2#Sec3)
  * [Array proliferation and interoperability](https://www.nature.com/articles/s41586-020-2649-2#Sec4)
  * [Discussion](https://www.nature.com/articles/s41586-020-2649-2#Sec5)
  * [References](https://www.nature.com/articles/s41586-020-2649-2#Bib1)
  * [Acknowledgements](https://www.nature.com/articles/s41586-020-2649-2#Ack1)
  * [Author information](https://www.nature.com/articles/s41586-020-2649-2#author-information)
  * [Ethics declarations](https://www.nature.com/articles/s41586-020-2649-2#ethics)
  * [Additional information](https://www.nature.com/articles/s41586-020-2649-2#additional-information)
  * [Supplementary information](https://www.nature.com/articles/s41586-020-2649-2#Sec7)
  * [Rights and permissions](https://www.nature.com/articles/s41586-020-2649-2#rightslink)
  * [About this article](https://www.nature.com/articles/s41586-020-2649-2#article-info)
  * [This article is cited by](https://www.nature.com/articles/s41586-020-2649-2#further-reading)


  * **Fig. 1: The NumPy array incorporates several fundamental array concepts.**
[View in article](https://www.nature.com/articles/s41586-020-2649-2#Fig1)[Full size image](https://www.nature.com/articles/s41586-020-2649-2/figures/1)
  * **Fig. 2: NumPy is the base of the scientific Python ecosystem.**
[View in article](https://www.nature.com/articles/s41586-020-2649-2#Fig2)[Full size image](https://www.nature.com/articles/s41586-020-2649-2/figures/2)
  * **Fig. 3: NumPy’s API and array protocols expose new arrays to the ecosystem.**
[View in article](https://www.nature.com/articles/s41586-020-2649-2#Fig3)[Full size image](https://www.nature.com/articles/s41586-020-2649-2/figures/3)


  1. Abbott, B. P. et al. Observation of gravitational waves from a binary black hole merger. _Phys. Rev. Lett_. **116** , 061102 (2016).
[Article](https://doi.org/10.1103%2FPhysRevLett.116.061102) [ADS](http://adsabs.harvard.edu/cgi-bin/nph-data_query?link_type=ABSTRACT&bibcode=2016PhRvL.116f1102A) [MathSciNet](http://www.ams.org/mathscinet-getitem?mr=3707758) [CAS](https://www.nature.com/articles/cas-redirect/1:STN:280:DC%2BC28jltVGjug%3D%3D) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Observation%20of%20gravitational%20waves%20from%20a%20binary%20black%20hole%20merger&journal=Phys.%20Rev.%20Lett.&doi=10.1103%2FPhysRevLett.116.061102&volume=116&publication_year=2016&author=Abbott%2CBP)
  2. Chael, A. et al. High-resolution linear polarimetric imaging for the Event Horizon Telescope. _Astrophys. J_. **286** , 11 (2016).
[Article](https://doi.org/10.3847%2F0004-637X%2F829%2F1%2F11) [ADS](http://adsabs.harvard.edu/cgi-bin/nph-data_query?link_type=ABSTRACT&bibcode=2016ApJ...833...11C) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=High-resolution%20linear%20polarimetric%20imaging%20for%20the%20Event%20Horizon%20Telescope&journal=Astrophys.%20J.&doi=10.3847%2F0004-637X%2F829%2F1%2F11&volume=286&publication_year=2016&author=Chael%2CA)
  3. Dubois, P. F., Hinsen, K. & Hugunin, J. Numerical Python. _Comput. Phys_. **10** , 262–267 (1996).
[Article](https://doi.org/10.1063%2F1.4822400) [ADS](http://adsabs.harvard.edu/cgi-bin/nph-data_query?link_type=ABSTRACT&bibcode=1996ComPh..10..262D) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Numerical%20Python&journal=Comput.%20Phys.&doi=10.1063%2F1.4822400&volume=10&pages=262-267&publication_year=1996&author=Dubois%2CPF&author=Hinsen%2CK&author=Hugunin%2CJ)
  4. Ascher, D., Dubois, P. F., Hinsen, K., Hugunin, J. & Oliphant, T. E. _An Open Source Project: Numerical Python_ (Lawrence Livermore National Laboratory, 2001).
  5. Yang, T.-Y., Furnish, G. & Dubois, P. F. Steering object-oriented scientific computations. In _Proc. TOOLS USA 97. Intl Conf. Technology of Object Oriented Systems and Languages_ (eds Ege, R., Singh, M. & Meyer, B.) 112–119 (IEEE, 1997).
  6. Greenfield, P., Miller, J. T., Hsu, J. & White, R. L. numarray: a new scientific array package for Python. In _PyCon DC 2003_ <http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.112.9899> (2003).
  7. Oliphant, T. E. _Guide to NumPy_ 1st edn (Trelgol Publishing, 2006).
  8. Dubois, P. F. Python: batteries included. _Comput. Sci. Eng_. **9** , 7–9 (2007).
[Article](https://doi.org/10.1109%2FMCSE.2007.51) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Python%3A%20batteries%20included&journal=Comput.%20Sci.%20Eng.&doi=10.1109%2FMCSE.2007.51&volume=9&pages=7-9&publication_year=2007&author=Dubois%2CPF)
  9. Oliphant, T. E. Python for scientific computing. _Comput. Sci. Eng_. **9** , 10–20 (2007).
[Article](https://doi.org/10.1109%2FMCSE.2007.58) [CAS](https://www.nature.com/articles/cas-redirect/1:CAS:528:DC%2BD2sXltVWiurw%3D) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Python%20for%20scientific%20computing&journal=Comput.%20Sci.%20Eng.&doi=10.1109%2FMCSE.2007.58&volume=9&pages=10-20&publication_year=2007&author=Oliphant%2CTE)
  10. Millman, K. J. & Aivazis, M. Python for scientists and engineers. _Comput. Sci. Eng_. **13** , 9–12 (2011).
[Article](https://doi.org/10.1109%2FMCSE.2011.36) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Python%20for%20scientists%20and%20engineers&journal=Comput.%20Sci.%20Eng.&doi=10.1109%2FMCSE.2011.36&volume=13&pages=9-12&publication_year=2011&author=Millman%2CKJ&author=Aivazis%2CM)
  11. Pérez, F., Granger, B. E. & Hunter, J. D. Python: an ecosystem for scientific computing. _Comput. Sci. Eng_. **13** , 13–21 (2011). **Explains why the scientific Python ecosystem is a highly productive environment for research**.
[Article](https://doi.org/10.1109%2FMCSE.2010.119) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Python%3A%20an%20ecosystem%20for%20scientific%20computing&journal=Comput.%20Sci.%20Eng.&doi=10.1109%2FMCSE.2010.119&volume=13&pages=13-21&publication_year=2011&author=P%C3%A9rez%2CF&author=Granger%2CBE&author=Hunter%2CJD)
  12. Virtanen, P. et al. SciPy 1.0—fundamental algorithms for scientific computing in Python. _Nat. Methods_ **17** , 261–272 (2020); correction **17** , 352 (2020). **Introduces the SciPy library and includes a more detailed history of NumPy and SciPy.**
[Article](https://doi.org/10.1038%2Fs41592-019-0686-2) [CAS](https://www.nature.com/articles/cas-redirect/1:CAS:528:DC%2BB3cXislCjuro%3D) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=SciPy%201.0%E2%80%94fundamental%20algorithms%20for%20scientific%20computing%20in%20Python&journal=Nat.%20Methods&doi=10.1038%2Fs41592-019-0686-2&volume=17&pages=261-272&publication_year=2020&author=Virtanen%2CP)
  13. Hunter, J. D. Matplotlib: a 2D graphics environment. _Comput. Sci. Eng_. **9** , 90–95 (2007).
[Article](https://doi.org/10.1109%2FMCSE.2007.55) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Matplotlib%3A%20a%202D%20graphics%20environment&journal=Comput.%20Sci.%20Eng.&doi=10.1109%2FMCSE.2007.55&volume=9&pages=90-95&publication_year=2007&author=Hunter%2CJD)
  14. McKinney, W. Data structures for statistical computing in Python. In _Proc. 9th Python in Science Conf_. (eds van der Walt, S. & Millman, K. J.) 56–61 (2010).
  15. Pedregosa, F. et al. Scikit-learn: machine learning in Python. _J. Mach. Learn. Res_. **12** , 2825–2830 (2011).
[MathSciNet](http://www.ams.org/mathscinet-getitem?mr=2854348) [MATH](http://www.emis.de/MATH-item?1280.68189) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Scikit-learn%3A%20machine%20learning%20in%20Python&journal=J.%20Mach.%20Learn.%20Res.&volume=12&pages=2825-2830&publication_year=2011&author=Pedregosa%2CF)
  16. van der Walt, S. et al. scikit-image: image processing in Python. _PeerJ_ **2** , e453 (2014).
[Article](https://doi.org/10.7717%2Fpeerj.453) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=scikit-image%3A%20image%20processing%20in%20Python&journal=PeerJ&doi=10.7717%2Fpeerj.453&volume=2&publication_year=2014&author=Walt%2CS)
  17. van der Walt, S., Colbert, S. C. & Varoquaux, G. The NumPy array: a structure for efficient numerical computation. _Comput. Sci. Eng_. **13** , 22–30 (2011). **Discusses the NumPy array data structure with a focus on how it enables efficient computation**.
  18. Wang, Q., Zhang, X., Zhang, Y. & Yi, Q. AUGEM: automatically generate high performance dense linear algebra kernels on x86 CPUs. In _SC’13: Proc. Intl Conf. High Performance Computing, Networking, Storage and Analysis_ 25 (IEEE, 2013).
  19. Xianyi, Z., Qian, W. & Yunquan, Z. Model-driven level 3 BLAS performance optimization on Loongson 3A processor. In _2012 IEEE 18th Intl Conf. Parallel and Distributed Systems_ 684–691 (IEEE, 2012).
  20. Pérez, F. & Granger, B. E. IPython: a system for interactive scientific computing. _Comput. Sci. Eng_. **9** , 21–29 (2007).
[Article](https://doi.org/10.1109%2FMCSE.2007.53) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=IPython%3A%20a%20system%20for%20interactive%20scientific%20computing&journal=Comput.%20Sci.%20Eng.&doi=10.1109%2FMCSE.2007.53&volume=9&pages=21-29&publication_year=2007&author=P%C3%A9rez%2CF&author=Granger%2CBE)
  21. Kluyver, T. et al. Jupyter Notebooks—a publishing format for reproducible computational workflows. In _Positioning and Power in Academic Publishing: Players, Agents and Agendas_ (eds Loizides, F. & Schmidt, B.) 87–90 (IOS Press, 2016).
  22. Hagberg, A. A., Schult, D. A. & Swart, P. J. Exploring network structure, dynamics, and function using NetworkX. In _Proc. 7th Python in Science Conf_. (eds Varoquaux, G., Vaught, T. & Millman, K. J.) 11–15 (2008).
  23. Astropy Collaboration et al. Astropy: a community Python package for astronomy. _Astron. Astrophys_. **558** , A33 (2013).
[Article](https://doi.org/10.1051%2F0004-6361%2F201322068) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Astropy%3A%20a%20community%20Python%20package%20for%20astronomy&journal=Astron.%20Astrophys.&doi=10.1051%2F0004-6361%2F201322068&volume=558&publication_year=2013&author=Astropy%20Collaboration%2C)
  24. Price-Whelan, A. M. et al. The Astropy Project: building an open-science project and status of the v2.0 core package. _Astron. J_. **156** , 123 (2018).
[Article](https://doi.org/10.3847%2F1538-3881%2Faac387) [ADS](http://adsabs.harvard.edu/cgi-bin/nph-data_query?link_type=ABSTRACT&bibcode=2018AJ....156..123A) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=The%20Astropy%20Project%3A%20building%20an%20open-science%20project%20and%20status%20of%20the%20v2.0%20core%20package&journal=Astron.%20J.&doi=10.3847%2F1538-3881%2Faac387&volume=156&publication_year=2018&author=Price-Whelan%2CAM)
  25. Cock, P. J. et al. Biopython: freely available Python tools for computational molecular biology and bioinformatics. _Bioinformatics_ **25** , 1422–1423 (2009).
[Article](https://doi.org/10.1093%2Fbioinformatics%2Fbtp163) [CAS](https://www.nature.com/articles/cas-redirect/1:CAS:528:DC%2BD1MXmtFeqt74%3D) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Biopython%3A%20freely%20available%20Python%20tools%20for%20computational%20molecular%20biology%20and%20bioinformatics&journal=Bioinformatics&doi=10.1093%2Fbioinformatics%2Fbtp163&volume=25&pages=1422-1423&publication_year=2009&author=Cock%2CPJ)
  26. Millman, K. J. & Brett, M. Analysis of functional magnetic resonance imaging in Python. _Comput. Sci. Eng_. **9** , 52–55 (2007).
[Article](https://doi.org/10.1109%2FMCSE.2007.46) [CAS](https://www.nature.com/articles/cas-redirect/1:CAS:528:DC%2BD2sXltVWiurs%3D) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Analysis%20of%20functional%20magnetic%20resonance%20imaging%20in%20Python&journal=Comput.%20Sci.%20Eng.&doi=10.1109%2FMCSE.2007.46&volume=9&pages=52-55&publication_year=2007&author=Millman%2CKJ&author=Brett%2CM)
  27. The SunPy Community et al. SunPy—Python for solar physics. _Comput. Sci. Discov_. **8** , 014009 (2015).
[Article](https://doi.org/10.1088%2F1749-4699%2F8%2F1%2F014009) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=SunPy%E2%80%94Python%20for%20solar%20physics&journal=Comput.%20Sci.%20Discov.&doi=10.1088%2F1749-4699%2F8%2F1%2F014009&volume=8&publication_year=2015)
  28. Hamman, J., Rocklin, M. & Abernathy, R. Pangeo: a big-data ecosystem for scalable Earth system science. In _EGU General Assembly Conf. Abstracts_ 12146 (2018).
  29. Chael, A. A. et al. ehtim: imaging, analysis, and simulation software for radio interferometry. _Astrophysics Source Code Library_ <https://ascl.net/1904.004> (2019).
  30. Millman, K. J. & Pérez, F. Developing open source scientific practice. In _Implementing Reproducible Research_ (eds Stodden, V., Leisch, F. & Peng, R. D.) 149–183 (CRC Press, 2014). **Describes the software engineering practices embraced by the NumPy and SciPy communities with a focus on how these practices improve research**.
  31. van der Walt, S. The SciPy Documentation Project (technical overview). In _Proc. 7th Python in Science Conf. (SciPy 2008)_ (eds Varoquaux, G., Vaught, T. & Millman, K. J.) 27–28 (2008).
  32. Harrington, J. The SciPy Documentation Project. In _Proc. 7th Python in Science Conference (SciPy 2008)_ (eds Varoquaux, G., Vaught, T. & Millman, K. J.) 33–35 (2008).
  33. Harrington, J. & Goldsmith, D. Progress report: NumPy and SciPy documentation in 2009. In _Proc. 8th Python in Science Conf. (SciPy 2009)_ (eds Varoquaux, G., van der Walt, S. & Millman, K. J.) 84–87 (2009).
  34. Royal Astronomical Society _Report of the RAS ‘A’ Awards Committee 2020: Astropy Project: 2020 Group Achievement Award (A)_ <https://ras.ac.uk/sites/default/files/2020-01/Group%20Award%20-%20Astropy.pdf> (2020).
  35. Wilson, G. Software carpentry: getting scientists to write better code by making them more productive. _Comput. Sci. Eng_. **8** , 66–69 (2006).
[Article](https://doi.org/10.1109%2FMCSE.2006.122) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Software%20carpentry%3A%20getting%20scientists%20to%20write%20better%20code%20by%20making%20them%20more%20productive&journal=Comput.%20Sci.%20Eng.&doi=10.1109%2FMCSE.2006.122&volume=8&pages=66-69&publication_year=2006&author=Wilson%2CG)
  36. Hannay, J. E. et al. How do scientists develop and use scientific software? In _Proc. 2009 ICSE Workshop on Software Engineering for Computational Science and Engineering_ 1–8 (IEEE, 2009).
  37. Millman, K. J., Brett, M., Barnowski, R. & Poline, J.-B. Teaching computational reproducibility for neuroimaging. _Front. Neurosci_. **12** , 727 (2018).
[Article](https://doi.org/10.3389%2Ffnins.2018.00727) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Teaching%20computational%20reproducibility%20for%20neuroimaging&journal=Front.%20Neurosci.&doi=10.3389%2Ffnins.2018.00727&volume=12&publication_year=2018&author=Millman%2CKJ&author=Brett%2CM&author=Barnowski%2CR&author=Poline%2CJ-B)
  38. Paszke, A. et al. Pytorch: an imperative style, high-performance deep learning library. In _Advances in Neural Information Processing Systems 32_ (eds Wallach, H. et al.) 8024–8035 (Neural Information Processing Systems, 2019).
  39. Abadi, M. et al. TensorFlow: a system for large-scale machine learning. In _OSDI’16: Proc. 12th USENIX Conf. Operating Systems Design and Implementation_ (chairs Keeton, K. & Roscoe, T.) 265–283 (USENIX Association, 2016).
  40. Chen, T. et al. MXNet: a flexible and efficient machine learning library for heterogeneous distributed systems. Preprint at <http://www.arxiv.org/abs/1512.01274> (2015).
  41. Hoyer, S. & Hamman, J. xarray: N–D labeled arrays and datasets in Python. _J. Open Res. Softw_. **5** , 10 (2017).
[Article](https://doi.org/10.5334%2Fjors.148) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=xarray%3A%20N%E2%80%93D%20labeled%20arrays%20and%20datasets%20in%20Python&journal=J.%20Open%20Res.%20Softw.&doi=10.5334%2Fjors.148&volume=5&publication_year=2017&author=Hoyer%2CS&author=Hamman%2CJ)
  42. Entschev, P. Distributed multi-GPU computing with Dask, CuPy and RAPIDS. In _EuroPython 2019_ <https://ep2019.europython.eu/media/conference/slides/fX8dJsD-distributed-multi-gpu-computing-with-dask-cupy-and-rapids.pdf> (2019).
  43. Behnel, S. et al. Cython: the best of both worlds. _Comput. Sci. Eng_. **13** , 31–39 (2011).
[Article](https://doi.org/10.1109%2FMCSE.2010.118) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Cython%3A%20the%20best%20of%20both%20worlds&journal=Comput.%20Sci.%20Eng.&doi=10.1109%2FMCSE.2010.118&volume=13&pages=31-39&publication_year=2011&author=Behnel%2CS)
  44. Lam, S. K., Pitrou, A. & Seibert, S. Numba: a LLVM-based Python JIT compiler. In _Proc. Second Workshop on the LLVM Compiler Infrastructure in HPC, LLVM ’15_ 7:1–7:6 (ACM, 2015).
  45. Guelton, S. et al. Pythran: enabling static optimization of scientific Python programs. _Comput. Sci. Discov_. **8** , 014001 (2015).
[Article](https://doi.org/10.1088%2F1749-4680%2F8%2F1%2F014001) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Pythran%3A%20enabling%20static%20optimization%20of%20scientific%20Python%20programs&journal=Comput.%20Sci.%20Discov.&doi=10.1088%2F1749-4680%2F8%2F1%2F014001&volume=8&publication_year=2015&author=Guelton%2CS)
  46. Dongarra, J., Golub, G. H., Grosse, E., Moler, C. & Moore, K. Netlib and NA-Net: building a scientific computing community. _IEEE Ann. Hist. Comput_. **30** , 30–41 (2008).
[Article](https://doi.org/10.1109%2FMAHC.2008.29) [MathSciNet](http://www.ams.org/mathscinet-getitem?mr=2432392) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Netlib%20and%20NA-Net%3A%20building%20a%20scientific%20computing%20community&journal=IEEE%20Ann.%20Hist.%20Comput.&doi=10.1109%2FMAHC.2008.29&volume=30&pages=30-41&publication_year=2008&author=Dongarra%2CJ&author=Golub%2CGH&author=Grosse%2CE&author=Moler%2CC&author=Moore%2CK)
  47. Barrett, K. A., Chiu, Y. H., Painter, J. F., Motteler, Z. C. & Dubois, P. F. _Basis System, Part I: Running a Basis Program—A Tutorial for Beginners_ UCRL-MA-118543, Vol. 1 (Lawrence Livermore National Laboratory 1995).
  48. Dubois, P. F. & Motteler, Z. _Basis System, Part II: Basis Language Reference Manual_ UCRL-MA-118543, Vol. 2 (Lawrence Livermore National Laboratory, 1995).
  49. Chiu, Y. H. & Dubois, P. F. _Basis System, Part III: EZN User Manual_ UCRL-MA-118543, Vol. 3 (Lawrence Livermore National Laboratory, 1995).
  50. Chiu, Y. H. & Dubois, P. F. _Basis System, Part IV: EZD User Manual_ UCRL-MA-118543, Vol. 4 (Lawrence Livermore National Laboratory, 1995).
  51. Munro, D. H. & Dubois, P. F. Using the Yorick interpreted language. _Comput. Phys_. **9** , 609–615 (1995).
[Article](https://doi.org/10.1063%2F1.4823451) [ADS](http://adsabs.harvard.edu/cgi-bin/nph-data_query?link_type=ABSTRACT&bibcode=1995ComPh...9..609M) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Using%20the%20Yorick%20interpreted%20language&journal=Comput.%20Phys.&doi=10.1063%2F1.4823451&volume=9&pages=609-615&publication_year=1995&author=Munro%2CDH&author=Dubois%2CPF)
  52. Ihaka, R. & Gentleman, R. R: a language for data analysis and graphics. _J. Comput. Graph. Stat_. **5** , 299–314 (1996).
[ Google Scholar](http://scholar.google.com/scholar_lookup?&title=R%3A%20a%20language%20for%20data%20analysis%20and%20graphics&journal=J.%20Comput.%20Graph.%20Stat.&volume=5&pages=299-314&publication_year=1996&author=Ihaka%2CR&author=Gentleman%2CR)
  53. Iverson, K. E. A programming language. In _Proc. 1962 Spring Joint Computer Conf_. 345–351 (1962).
  54. Jenness, T. et al. LSST data management software development practices and tools. In _Proc. SPIE 10707, Software and Cyberinfrastructure for Astronomy V_ 1070709 (SPIE and International Society for Optics and Photonics, 2018).
  55. Matsakis, N. D. & Klock, F. S. The Rust language. _Ada Letters_ **34** , 103–104 (2014).
[Article](https://doi.org/10.1145%2F2692956.2663188) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=The%20Rust%20language&journal=Ada%20Letters&doi=10.1145%2F2692956.2663188&volume=34&pages=103-104&publication_year=2014&author=Matsakis%2CND&author=Klock%2CFS)
  56. Bezanson, J., Edelman, A., Karpinski, S. & Shah, V. B. Julia: a fresh approach to numerical computing. _SIAM Rev_. **59** , 65–98 (2017).
[Article](https://doi.org/10.1137%2F141000671) [MathSciNet](http://www.ams.org/mathscinet-getitem?mr=3605826) [ Google Scholar](http://scholar.google.com/scholar_lookup?&title=Julia%3A%20a%20fresh%20approach%20to%20numerical%20computing&journal=SIAM%20Rev.&doi=10.1137%2F141000671&volume=59&pages=65-98&publication_year=2017&author=Bezanson%2CJ&author=Edelman%2CA&author=Karpinski%2CS&author=Shah%2CVB)
  57. Lattner, C. & Adve, V. LLVM: a compilation framework for lifelong program analysis and transformation. In _Proc. 2004 Intl Symp. Code Generation and Optimization (CGO’04)_ 75–88 (IEEE, 2004).


Nature  (_Nature_) 
ISSN 1476-4687 (online) 
ISSN 0028-0836 (print) 
### About Nature Portfolio
  * [About us](https://www.nature.com/npg_/company_info/index.html)
  * [Press releases](https://www.nature.com/npg_/press_room/press_releases.html)
  * [Press office](https://press.nature.com/)
  * [Contact us](https://support.nature.com/support/home)


### Discover content
  * [Journals A-Z](https://www.nature.com/siteindex)
  * [Articles by subject](https://www.nature.com/subjects)
  * [protocols.io](https://www.protocols.io/)
  * [Nature Index](https://www.natureindex.com/)


### Publishing policies
  * [Nature portfolio policies](https://www.nature.com/authors/editorial_policies)
  * [Open access](https://www.nature.com/nature-research/open-access)


### Author & Researcher services
  * [Reprints & permissions](https://www.nature.com/reprints)
  * [Research data](https://www.springernature.com/gp/authors/research-data)
  * [Language editing](https://authorservices.springernature.com/language-editing/)
  * [Scientific editing](https://authorservices.springernature.com/scientific-editing/)
  * [Nature Masterclasses](https://masterclasses.nature.com/)
  * [Research Solutions](https://solutions.springernature.com/)


### Libraries & institutions
  * [Librarian service & tools](https://www.springernature.com/gp/librarians/tools-services)
  * [Librarian portal](https://www.springernature.com/gp/librarians/manage-your-account/librarianportal)
  * [Open research](https://www.nature.com/openresearch/about-open-access/information-for-institutions)
  * [Recommend to library](https://www.springernature.com/gp/librarians/recommend-to-your-library)


### Advertising & partnerships
  * [Advertising](https://partnerships.nature.com/product/digital-advertising/)
  * [Partnerships & Services](https://partnerships.nature.com/)
  * [Media kits](https://partnerships.nature.com/media-kits/)
  * [Branded content](https://partnerships.nature.com/product/branded-content-native-advertising/)


### Professional development
  * [Nature Awards](https://www.nature.com/immersive/natureawards/index.html)
  * [Nature Careers](https://www.nature.com/naturecareers/)
  * [Nature Conferences](https://conferences.nature.com)


### Regional websites
  * [Nature Africa](https://www.nature.com/natafrica)
  * [Nature China](http://www.naturechina.com)
  * [Nature India](https://www.nature.com/nindia)
  * [Nature Japan](https://www.natureasia.com/ja-jp)
  * [Nature Middle East](https://www.nature.com/nmiddleeast)


  * [Privacy Policy](https://www.nature.com/info/privacy)
  * [Use of cookies](https://www.nature.com/info/cookies)
  * Your privacy choices/Manage cookies 
  * [Legal notice](https://www.nature.com/info/legal-notice)
  * [Accessibility statement](https://www.nature.com/info/accessibility-statement)
  * [Terms & Conditions](https://www.nature.com/info/terms-and-conditions)
  * [Your US state privacy rights](https://www.springernature.com/ccpa)


[ ![Springer Nature](https://www.nature.com/static/images/logos/sn-logo-white-ea63208b81.svg) ](https://www.springernature.com/)
© 2025 Springer Nature Limited
Close banner Close
![Nature Briefing AI and Robotics](https://www.nature.com/static/images/logos/nature-briefing-ai-and-robotics-logo-51b3cf6c52.svg)
Sign up for the _Nature Briefing: AI and Robotics_ newsletter — what matters in AI and robotics research, free to your inbox weekly.
Email address
Sign up
I agree my information will be processed in accordance with the _Nature_ and Springer Nature Limited [Privacy Policy](https://www.nature.com/info/privacy).
Close banner Close
Get the most important science stories of the day, free in your inbox. [Sign up for Nature Briefing: AI and Robotics ](https://www.nature.com/briefing/ai-and-robotics/?brieferEntryPoint=AIAndRoboticsBriefingBanner)
  *[DOI]: Digital Object Identifier
  *[ISSN]: International Standard Serial Number
